{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57ebbf6d",
   "metadata": {},
   "source": [
    "#### This particular notebook details the combined preprocessing pipeline for our Xenium Dataset 1 (integrated rep 1 and 2) data\n",
    "\n",
    "#### Required input files:\n",
    "\n",
    "* Raw Xenium Dataset 1 290 IntReps1and2 data object (availabile via FigShare) OR raw data files for Xenium Dataset 1 290 Rep1 and Xenium Dataset 1 290 Rep2 (available via GEO)\n",
    "\n",
    "This notebook starts with the raw GEO files\n",
    "\n",
    "Note: Rep1 is occasionally referred to as the 'Sept' run, while Rep2 is occasionally referred to as the 'Nov' run.\n",
    "\n",
    "Environment: Please create and activate the conda environment provided in harmony_integration_env.yaml before running this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd1e7b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "\n",
    "import gzip\n",
    "import anndata\n",
    "\n",
    "# New addition for harmony_integration_env.yaml compared to default_env.yaml\n",
    "import harmonypy as hm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2b37fb9",
   "metadata": {},
   "source": [
    "### Load Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "293b9c2f",
   "metadata": {},
   "source": [
    "### Nov / Rep2 run"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f20ad45b",
   "metadata": {},
   "source": [
    "Load in essential data object files and store df within the object obs slot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e76c2a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov = sc.read_10x_h5(\n",
    "    filename=\"/path/cell_feature_matrix.h5\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72174f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can unzip cells.csz.gz to obtain cells.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a820725",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Nov = pd.read_csv(\n",
    "    \"/path/cells.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d41a88e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Nov.set_index(adata_Nov.obs_names, inplace=True)\n",
    "adata_Nov.obs = df_Nov.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ebc97e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store coordinates within obsm spatial\n",
    "adata_Nov.obsm[\"spatial\"] = adata_Nov.obs[[\"x_centroid\", \"y_centroid\"]].copy().to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eed35a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov.obsm[\"spatial\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38c19290",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov.obs\n",
    "\n",
    "# 424,230 cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eca8ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f00f59c9",
   "metadata": {},
   "source": [
    "#### Extra data object edits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "691fddd4",
   "metadata": {},
   "source": [
    "Rename a few metadata column names (to match previous notebooks) and add a column for batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a7dba65",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov.obs.rename(columns={'cell_area': 'cell_area_um2'}, inplace=True)\n",
    "\n",
    "adata_Nov.obs['batch'] = 'Nov'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da09c791",
   "metadata": {},
   "source": [
    "Continue -- Add cell_id back as row names and as an individual column. \n",
    "\n",
    "Additionally, rename each cell_id value to start with \"Nov-\" to differentiate the Xenium runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b876dbae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename the cell_id values to start with \"Nov-\" to denote the appropriate Xenium run\n",
    "adata_Nov_metadata[\"cell_id\"] = \"Nov-\" + adata_Nov_metadata[\"cell_id\"]\n",
    "\n",
    "\n",
    "### Add the cell_ids as row names, but keep as a column\n",
    "\n",
    "# Copy 'cell_id' to a new column\n",
    "adata_Nov_metadata['cell_id2'] = adata_Nov_metadata['cell_id']\n",
    "\n",
    "# Set 'cell_id2' as the index\n",
    "adata_Nov_metadata.set_index('cell_id2', inplace=True)\n",
    "\n",
    "# Remove the name of the index\n",
    "adata_Nov_metadata.index.name = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9be69b3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the modified metadata back into the adata_Nov object and view\n",
    "\n",
    "adata_Nov.obs = adata_Nov_metadata\n",
    "\n",
    "adata_Nov.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "576e603b",
   "metadata": {},
   "source": [
    "Offset x_centroid (just for 1 of the 2 datasets (choosing Nov run) so that we can plot them in the same space) and store coordinates (centroid x and y values) within the object obsm spatial slot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d8c586",
   "metadata": {},
   "outputs": [],
   "source": [
    "# UPDATE -- Adding / offsetting x_centroid by 15,000 (since adata_Sept.obs['x_centroid'].max() = ~10,000)\n",
    "adata_Nov.obs['x_centroid'] += 15000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25abf0f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store UPDATED coordinates within obsm spatial\n",
    "adata_Nov.obsm[\"spatial\"] = adata_Nov.obs[[\"x_centroid\", \"y_centroid\"]].copy().to_numpy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5ca7047",
   "metadata": {},
   "source": [
    "View the newly formed and edited data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "790eef89",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov.obs\n",
    "\n",
    "# 424,230 cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9458383",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3cdd799",
   "metadata": {},
   "source": [
    "### Sept / Rep1 run"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d26353e5",
   "metadata": {},
   "source": [
    "Load in essential data object files and store df within the object obs slot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e4508f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Sept = sc.read_10x_h5(\n",
    "    filename=\"/path/cell_feature_matrix.h5\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1108636",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We need to unzip cells.csz.gz to obtain cells.csv -- Already done, so I'm commenting out / deleting this section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecf83bd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Sept = pd.read_csv(\n",
    "    \"/path/cells.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0d673b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Sept.set_index(adata_Sept.obs_names, inplace=True)\n",
    "adata_Sept.obs = df_Sept.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85d25a50",
   "metadata": {},
   "source": [
    "View the newly formed and edited data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c8339b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Sept.obs\n",
    "\n",
    "# 434,666 cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25cda3ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Sept"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba3e4b6",
   "metadata": {},
   "source": [
    "#### Extra data object edits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3d7dd89",
   "metadata": {},
   "source": [
    "Rename a few metadata column names (to match previous notebooks) and add a column for batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bdba6df",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Sept.obs.rename(columns={'cell_area': 'cell_area_um2'}, inplace=True)\n",
    "\n",
    "adata_Sept.obs['batch'] = 'Sept'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9739bc0",
   "metadata": {},
   "source": [
    "Continue -- Add cell_id back as row names and as an individual column. \n",
    "\n",
    "Additionally, rename each cell_id value to start with \"Sept-\" to differentiate the Xenium runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d198fc9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename the cell_id values to start with \"Sept-\" to denote the appropriate Xenium run\n",
    "adata_Sept_metadata[\"cell_id\"] = \"Sept-\" + adata_Sept_metadata[\"cell_id\"]\n",
    "\n",
    "\n",
    "### Add the cell_ids as row names, but keep as a column\n",
    "\n",
    "# Copy 'cell_id' to a new column\n",
    "adata_Sept_metadata['cell_id2'] = adata_Sept_metadata['cell_id']\n",
    "\n",
    "# Set 'cell_id2' as the index\n",
    "adata_Sept_metadata.set_index('cell_id2', inplace=True)\n",
    "\n",
    "# Remove the name of the index\n",
    "adata_Sept_metadata.index.name = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6087af55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the modified metadata back into the adata_Sept object and view\n",
    "\n",
    "adata_Sept.obs = adata_Sept_metadata\n",
    "\n",
    "adata_Sept.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af5bbcf",
   "metadata": {},
   "source": [
    "Note: No need to offset x_centroid here since we already did it for the Nov run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "572ebd16",
   "metadata": {},
   "source": [
    "View the newly formed and edited data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18a2289a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Sept.obs\n",
    "\n",
    "# 434,666 cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "600cceca",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Sept"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aed76db7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata_Sept.obs['x_centroid'].max()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37a3071d",
   "metadata": {},
   "source": [
    "#### Concluding Notes (to inform concatenating)\n",
    "* All of the columns within the obs slot are the same. The row names (cell_id) have overlap -- but we've added identifier strings (Nov- or Sept-) to make them unique\n",
    "* Obsm spatial has the coordinates stored in an array -- different but overlapping. We've offset the x_centroid values for adata_Nov by 15,000\n",
    "* Var gene_ids, feature_types, and genome are the same across objects, since the same gene panel was used"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14375987",
   "metadata": {},
   "source": [
    "### Save data objects before continuing\n",
    "\n",
    "These adata objects can be loaded in for future versions of this pipeline. Note: They should only be used for this concat pipeline, due to particular formatting decisions (like adding the batch information and offsetting some spatial coordinates)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfd620b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov.write_h5ad('/path/XeniumNovData_initialformatforconcat.h5ad')\n",
    "adata_Sept.write_h5ad('/path/XeniumSeptData_initialformatforconcat.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6903d6af",
   "metadata": {},
   "source": [
    "### Calculate Quality Control Metrics (for both data objects)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef5f344f",
   "metadata": {},
   "source": [
    "Calculate the quality control metrics on the anndata.AnnData using scanpy.pp.calculate_qc_metrics\n",
    "\n",
    "Reminder -- This overrides Xeniun's original total_counts value, but I'm choosing to let this happen. The original counts value represents gene expression features, control probes, control codewords, and unassigned codewords. The new version includes (solely) the gene expression feature count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "919da246",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(adata_Sept, percent_top=(10, 20, 50, 150), inplace=True)\n",
    "\n",
    "sc.pp.calculate_qc_metrics(adata_Nov, percent_top=(10, 20, 50, 150), inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "658e2ad7",
   "metadata": {},
   "source": [
    "The percentage of control probes and control codewords can be calculated from adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d04b4f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cprobes_Sept = (\n",
    "    adata_Sept.obs[\"control_probe_counts\"].sum() / adata_Sept.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "cwords_Sept = (\n",
    "    adata_Sept.obs[\"control_codeword_counts\"].sum() / adata_Sept.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "\n",
    "print(\"adata_Sept\")\n",
    "print(f\"Negative DNA probe count % : {cprobes_Sept}\")\n",
    "print(f\"Negative decoding count % : {cwords_Sept}\")\n",
    "print(\"\\n\")\n",
    "\n",
    "cprobes_Nov = (\n",
    "    adata_Nov.obs[\"control_probe_counts\"].sum() / adata_Nov.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "cwords_Nov = (\n",
    "    adata_Nov.obs[\"control_codeword_counts\"].sum() / adata_Nov.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "\n",
    "print(\"adata_Nov\")\n",
    "print(f\"Negative DNA probe count % : {cprobes_Nov}\")\n",
    "print(f\"Negative decoding count % : {cwords_Nov}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5760b0d0",
   "metadata": {},
   "source": [
    "Next we plot the distribution of total transcripts per cell, unique transcripts per cell, area of segmented cells and the ratio of nuclei area to their cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d57a586a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sept\n",
    "\n",
    "fig, axs = plt.subplots(1, 4, figsize=(15, 4))\n",
    "\n",
    "axs[0].set_title(\"Total transcripts per cell\")\n",
    "sns.histplot(\n",
    "    adata_Sept.obs[\"total_counts\"],\n",
    "    kde=False,\n",
    "    ax=axs[0],\n",
    ")\n",
    "\n",
    "axs[1].set_title(\"Unique transcripts per cell\")\n",
    "sns.histplot(\n",
    "    adata_Sept.obs[\"n_genes_by_counts\"],\n",
    "    kde=False,\n",
    "    ax=axs[1],\n",
    ")\n",
    "\n",
    "\n",
    "axs[2].set_title(\"Area of segmented cells\")\n",
    "sns.histplot(\n",
    "    adata_Sept.obs[\"cell_area_um2\"],\n",
    "    kde=False,\n",
    "    ax=axs[2],\n",
    ")\n",
    "\n",
    "axs[3].set_title(\"Nucleus ratio\")\n",
    "sns.histplot(\n",
    "    adata_Sept.obs[\"nucleus_area\"] / adata_Sept.obs[\"cell_area_um2\"],\n",
    "    kde=False,\n",
    "    ax=axs[3],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d9a1c33",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Nov\n",
    "\n",
    "fig, axs = plt.subplots(1, 4, figsize=(15, 4))\n",
    "\n",
    "axs[0].set_title(\"Total transcripts per cell\")\n",
    "sns.histplot(\n",
    "    adata_Nov.obs[\"total_counts\"],\n",
    "    kde=False,\n",
    "    ax=axs[0],\n",
    ")\n",
    "\n",
    "axs[1].set_title(\"Unique transcripts per cell\")\n",
    "sns.histplot(\n",
    "    adata_Nov.obs[\"n_genes_by_counts\"],\n",
    "    kde=False,\n",
    "    ax=axs[1],\n",
    ")\n",
    "\n",
    "\n",
    "axs[2].set_title(\"Area of segmented cells\")\n",
    "sns.histplot(\n",
    "    adata_Nov.obs[\"cell_area_um2\"],\n",
    "    kde=False,\n",
    "    ax=axs[2],\n",
    ")\n",
    "\n",
    "axs[3].set_title(\"Nucleus ratio\")\n",
    "sns.histplot(\n",
    "    adata_Nov.obs[\"nucleus_area\"] / adata_Nov.obs[\"cell_area_um2\"],\n",
    "    kde=False,\n",
    "    ax=axs[3],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e21bfa7c",
   "metadata": {},
   "source": [
    "Filter the cells based on the minimum number of counts required using scanpy.pp.filter_cells. Filter the genes based on the minimum number of cells required with scanpy.pp.filter_genes. The parameters for the both were specified based on the plots above. They were set to filter out the cells and genes with minimum counts and minimum cells respectively.\n",
    "\n",
    "Other filter criteria might be cell area, DAPI signal or a minimum of unique transcripts.\n",
    "\n",
    "Squidpy tutorial filtering examples: \n",
    "sc.pp.filter_cells(adata, min_counts=10);\n",
    "sc.pp.filter_genes(adata, min_cells=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d64119c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Our filtering schema\n",
    "# Note: Transcripts that didn't pass 10x Genomics's QC have already been filtered out\n",
    "\n",
    "## Sept\n",
    "adata_Sept_filtered = adata_Sept\n",
    "\n",
    "# filter out cells with <50 counts and <10 genes\n",
    "sc.pp.filter_cells(adata_Sept_filtered, min_counts=50)\n",
    "sc.pp.filter_cells(adata_Sept_filtered, min_genes=10)\n",
    "\n",
    "# filter out genes that have <1 count and are detected in <10 cells\n",
    "sc.pp.filter_genes(adata_Sept_filtered, min_counts=1)\n",
    "sc.pp.filter_genes(adata_Sept_filtered, min_cells=10)\n",
    "\n",
    "\n",
    "## Nov\n",
    "adata_Nov_filtered = adata_Nov\n",
    "\n",
    "# filter out cells with <50 counts and <10 genes\n",
    "sc.pp.filter_cells(adata_Nov_filtered, min_counts=50)\n",
    "sc.pp.filter_cells(adata_Nov_filtered, min_genes=10)\n",
    "\n",
    "# filter out genes that have <1 count and are detected in <10 cells\n",
    "sc.pp.filter_genes(adata_Nov_filtered, min_counts=1)\n",
    "sc.pp.filter_genes(adata_Nov_filtered, min_cells=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f1e0102",
   "metadata": {},
   "source": [
    "#### Compare filtered objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cabeba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Sept_filtered.obs\n",
    "\n",
    "# 313,940 total cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6a2d4a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Nov_filtered.obs\n",
    "\n",
    "# 268,248 total cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3e78014",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nov -- 424,230 cells\n",
    "\n",
    "# Sept -- 434,666 cells\n",
    "\n",
    "\n",
    "## Total cells retained post-filtering (from 10x initial filtering step to our filtering step)\n",
    "# Values pasted from above\n",
    "\n",
    "print(f\"Sept % of cells maintained post my filtering: {(313940 / 434666) * 100}\")\n",
    "\n",
    "print(f\"Nov % of cells maintained post my filtering: {(268248 / 424230) * 100}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1012ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Total transcripts retained post-filtering (from 10x initial filtering step to our filtering step)\n",
    "\n",
    "total_counts_sum_Sept = adata_Sept.obs['total_counts'].sum()\n",
    "total_counts_sum_Sept_filtered = adata_Sept_filtered.obs['total_counts'].sum()\n",
    "\n",
    "percent_maintained_Sept = ((total_counts_sum_Sept_filtered / total_counts_sum_Sept) * 100)\n",
    "\n",
    "print(\"Sept % of transcripts (total_counts) maintained post my filtering:\", percent_maintained_Sept)\n",
    "\n",
    "total_counts_sum_Nov = adata_Nov.obs['total_counts'].sum()\n",
    "total_counts_sum_Nov_filtered = adata_Nov_filtered.obs['total_counts'].sum()\n",
    "\n",
    "percent_maintained_Nov = ((total_counts_sum_Nov_filtered / total_counts_sum_Nov) * 100)\n",
    "\n",
    "print(\"Nov % of transcripts (total_counts) maintained post my filtering:\", percent_maintained_Nov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f06b7009",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate min, mean, and max values of adata_Sept_filtered\n",
    "\n",
    "total_counts_minvalue = adata_Sept_filtered.obs[\"total_counts\"].min()\n",
    "total_counts_mean = adata_Sept_filtered.obs[\"total_counts\"].mean()\n",
    "total_counts_maxvalue = adata_Sept_filtered.obs[\"total_counts\"].max()\n",
    "\n",
    "n_genes_by_counts_minvalue = adata_Sept_filtered.obs[\"n_genes_by_counts\"].min()\n",
    "n_genes_by_counts_mean = adata_Sept_filtered.obs[\"n_genes_by_counts\"].mean()\n",
    "n_genes_by_counts_maxvalue = adata_Sept_filtered.obs[\"n_genes_by_counts\"].max()\n",
    "\n",
    "cell_area_um2_minvalue = adata_Sept_filtered.obs[\"cell_area_um2\"].min()\n",
    "cell_area_um2_mean = adata_Sept_filtered.obs[\"cell_area_um2\"].mean()\n",
    "cell_area_um2_maxvalue = adata_Sept_filtered.obs[\"cell_area_um2\"].max()\n",
    "\n",
    "print(\"adata_Sept_filtered\\n\")\n",
    "\n",
    "print(f\"Min value of total_counts : {total_counts_minvalue}\")\n",
    "print(f\"Mean of total_counts : {total_counts_mean}\")\n",
    "print(f\"Max value of total_counts : {total_counts_maxvalue}\\n\")\n",
    "\n",
    "print(f\"Min value of n_genes_by_counts : {n_genes_by_counts_minvalue}\")\n",
    "print(f\"Mean of n_genes_by_counts : {n_genes_by_counts_mean}\")\n",
    "print(f\"Max value of n_genes_by_counts : {n_genes_by_counts_maxvalue}\\n\")\n",
    "\n",
    "print(f\"Min value of cell area (um2) : {cell_area_um2_minvalue}\")\n",
    "print(f\"Mean of cell area (um2) : {cell_area_um2_mean}\")\n",
    "print(f\"Max value of cell area (um2) : {cell_area_um2_maxvalue}\\n\")\n",
    "\n",
    "## Calculate min, mean, and max values of adata_Nov_filtered\n",
    "\n",
    "total_counts_minvalue = adata_Nov_filtered.obs[\"total_counts\"].min()\n",
    "total_counts_mean = adata_Nov_filtered.obs[\"total_counts\"].mean()\n",
    "total_counts_maxvalue = adata_Nov_filtered.obs[\"total_counts\"].max()\n",
    "\n",
    "n_genes_by_counts_minvalue = adata_Nov_filtered.obs[\"n_genes_by_counts\"].min()\n",
    "n_genes_by_counts_mean = adata_Nov_filtered.obs[\"n_genes_by_counts\"].mean()\n",
    "n_genes_by_counts_maxvalue = adata_Nov_filtered.obs[\"n_genes_by_counts\"].max()\n",
    "\n",
    "cell_area_um2_minvalue = adata_Nov_filtered.obs[\"cell_area_um2\"].min()\n",
    "cell_area_um2_mean = adata_Nov_filtered.obs[\"cell_area_um2\"].mean()\n",
    "cell_area_um2_maxvalue = adata_Nov_filtered.obs[\"cell_area_um2\"].max()\n",
    "\n",
    "print(\"adata_Nov_filtered\\n\")\n",
    "\n",
    "print(f\"Min value of total_counts : {total_counts_minvalue}\")\n",
    "print(f\"Mean of total_counts : {total_counts_mean}\")\n",
    "print(f\"Max value of total_counts : {total_counts_maxvalue}\\n\")\n",
    "\n",
    "print(f\"Min value of n_genes_by_counts : {n_genes_by_counts_minvalue}\")\n",
    "print(f\"Mean of n_genes_by_counts : {n_genes_by_counts_mean}\")\n",
    "print(f\"Max value of n_genes_by_counts : {n_genes_by_counts_maxvalue}\\n\")\n",
    "\n",
    "print(f\"Min value of cell area (um2) : {cell_area_um2_minvalue}\")\n",
    "print(f\"Mean of cell area (um2) : {cell_area_um2_mean}\")\n",
    "print(f\"Max value of cell area (um2) : {cell_area_um2_maxvalue}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0477b053",
   "metadata": {},
   "source": [
    "Visualize genes with the highest expression levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5566aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highest_expr_genes(adata_Sept_filtered, n_top=20, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a35c0702",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highest_expr_genes(adata_Nov_filtered, n_top=20, )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a27c5650",
   "metadata": {},
   "source": [
    "### Concatenate the 2 objects into 1 anndata object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d2411ea",
   "metadata": {},
   "source": [
    "With concat(), AnnData objects can be combined via a composition of two operations: concatenation and merging.\n",
    "\n",
    "* Concatenation is when we keep all sub elements of each object, and stack these elements in an ordered way.\n",
    "* Merging is combining a set of collections into one resulting collection which contains elements from the objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c835e794",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate along the observations axis (axis=0)\n",
    "concat_filtered_ad = anndata.concat([adata_Sept_filtered, adata_Nov_filtered], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ce60b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfb4f365",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad.obs\n",
    "\n",
    "# 582,188 cells (313,940 from Sept and 268,248 from Nov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dfb3c62",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad.obsm['spatial']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c3b9d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.scatter(concat_filtered_ad, x='x_centroid', y='y_centroid', color='batch', show=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea41a9ba",
   "metadata": {},
   "source": [
    "### Now continue with analysis on concatenated object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2358a5d8",
   "metadata": {},
   "source": [
    "Make a copy of the original raw counts (post-filtering; pre-normalization)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a13e2551",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad.layers['raw_counts'] = concat_filtered_ad.X.copy() # Make a copy\n",
    "concat_filtered_ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce19e63c",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12d567a6",
   "metadata": {},
   "source": [
    "### Save object\n",
    "\n",
    "This is post-filtering and pre-normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22a97412",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad.write_h5ad('/path/DataObjects_withoutUMAP/Xenium_concatSeptNov_filtered_pre-normalization.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cacfd74",
   "metadata": {},
   "source": [
    "### Continue with analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b268c1de",
   "metadata": {},
   "source": [
    "Normalize counts per cell using scanpy.pp.normalize_total.\n",
    "\n",
    "Logarithmize, do principal component analysis, compute a neighborhood graph of the observations using scanpy.pp.log1p, scanpy.pp.pca and scanpy.pp.neighbors respectively.\n",
    "\n",
    "Use scanpy.tl.umap to embed the neighborhood graph of the data and cluster the cells into subgroups employing scanpy.tl.leiden."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c224f22",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(concat_filtered_ad, inplace=True)\n",
    "sc.pp.log1p(concat_filtered_ad)\n",
    "# Save log_normalized_counts as a layer\n",
    "concat_filtered_ad.layers['log_normalized_counts']=concat_filtered_ad.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a81450",
   "metadata": {},
   "source": [
    "Calculate and plot the top highly variable genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3bc8616",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(concat_filtered_ad, min_mean=0.0125, max_mean=3, min_disp=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0d56f88",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highly_variable_genes(concat_filtered_ad)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb7ec23f",
   "metadata": {},
   "source": [
    "Regress out unwanted sources or variation and scale data\n",
    "\n",
    "#### Note: All objects in this notebook will undergo this step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "258d4502",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here, we're just going to regress out total_counts and n_genes_by_counts\n",
    "\n",
    "sc.pp.regress_out(concat_filtered_ad, [\"total_counts\",\"n_genes_by_counts\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ae2b790",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.scale(concat_filtered_ad, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4de911e3",
   "metadata": {},
   "source": [
    "Run PCA and plot PCA variance ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f64ebf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.pca(concat_filtered_ad, svd_solver='arpack')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "370416a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca(concat_filtered_ad, color=['EPCAM','COL3A1','CD69'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c2ef334",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the PCA variance ratio\n",
    "sc.pl.pca_variance_ratio(concat_filtered_ad, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ea32d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the PCA variance ratio\n",
    "sc.pl.pca_variance_ratio(concat_filtered_ad, n_pcs = 50, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41796c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfc1c20d",
   "metadata": {},
   "source": [
    "### Save object\n",
    "\n",
    "This is post-filtering and post-normalization. Immediately before computing neighbor graph and plotting UMAP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b033e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad.write_h5ad('/path/DataObjects_withoutUMAP/Xenium_concatSeptNov_filtered_post-normalization.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b55c8f50",
   "metadata": {},
   "source": [
    "### Run integration with harmony\n",
    "\n",
    "(Even though these datasets are replicates, it's still good to run Harmony just in case)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac3cdfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare metadata and PCA\n",
    "\n",
    "meta_data = concat_filtered_ad.obs\n",
    "data_mat = concat_filtered_ad.obsm['X_pca']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e4f7fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run harmony\n",
    "\n",
    "ho = hm.run_harmony(data_mat, meta_data, 'batch')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73086fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping back the result to the adata object\n",
    "\n",
    "concat_filtered_ad.obsm['X_pca'] = ho.Z_corr.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b79b3b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save object\n",
    "\n",
    "concat_filtered_ad.write_h5ad('/path/DataObjects_withoutUMAP/Xenium_concatSeptNov_filtered_post-normalization_post-harmony.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9049f02b",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9efb8055",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad.obsm['X_pca']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6d591f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-plot PCA visualization\n",
    "sc.pl.pca(concat_filtered_ad, color=['EPCAM','COL3A1','CD69'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66796edc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-plot the second PCA variance ratio\n",
    "sc.pl.pca_variance_ratio(concat_filtered_ad, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4cb7b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-plot the second PCA variance ratio\n",
    "sc.pl.pca_variance_ratio(concat_filtered_ad, n_pcs= 50, log=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48f24143",
   "metadata": {},
   "source": [
    "### Compute neighbor graph and plot UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edbf66c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.figdir = '/path/UMAP_pngs/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae30692f",
   "metadata": {},
   "source": [
    "#### Parameter descriptions:\n",
    "n_neighbors\n",
    "* A value between 2 and 100, representing the number of neighboring data points used for manifld approximation. Larger values give a manifold with a more global view of the dataset, while smaller values preserve more of the local structures.\n",
    "* Default value is 15\n",
    "\n",
    "n_pcs\n",
    "* Use this many PCs\n",
    "* Default value is None\n",
    "* Choose the value after the elbow, not before\n",
    "\n",
    "min_dist\n",
    "* The minimum distance between two points in the UMAP embedding.\n",
    "* Default value is 0.05\n",
    "\n",
    "spread\n",
    "* A scaling factor for distance between embedded points.\n",
    "* Default value is 1.0\n",
    "\n",
    "Helpful resource: https://smorabit.github.io/blog/2020/umap/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a40a34c",
   "metadata": {},
   "source": [
    "#### Testing v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4d21f56",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_filtered_ad_v2 = concat_filtered_ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e0c0379",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(concat_filtered_ad_v2, n_neighbors=10, n_pcs=30)\n",
    "sc.tl.umap(concat_filtered_ad_v2, min_dist=0.02, spread=1.5)\n",
    "sc.tl.leiden(concat_filtered_ad_v2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7353e905",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    concat_filtered_ad_v2,\n",
    "    color=[\n",
    "        \"total_counts\",\n",
    "        \"n_genes_by_counts\",\n",
    "        \"leiden\",\n",
    "    ],\n",
    "    wspace=0.4,\n",
    "    save = '_v2_Xeniumconcat.png',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb489d5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    concat_filtered_ad_v2,\n",
    "    color=\"batch\",\n",
    "    wspace=0.4,\n",
    "    save=\"_v2_Xeniumconcat_splitbybatch.png\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f465452",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save object with UMAP\n",
    "concat_filtered_ad_v2.write_h5ad('/path/DataObjects_withUMAP/Xeniumconcat_umapv2_231210.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "432a22b1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "927ef967",
   "metadata": {},
   "source": [
    "### Generate umap without running harmony (for batch effect reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42e7a644",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in file (no harmony v2 umap)\n",
    "\n",
    "concat_filtered_ad_nhv2 = sc.read_h5ad('/path/DataObjects_withoutUMAP/Xenium_concatSeptNov_filtered_post-normalization_231210.h5ad')\n",
    "\n",
    "# View\n",
    "concat_filtered_ad_nhv2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6490f906",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(concat_filtered_ad_nhv2, n_neighbors=10, n_pcs=30)\n",
    "sc.tl.umap(concat_filtered_ad_nhv2, min_dist=0.02, spread=1.5)\n",
    "sc.tl.leiden(concat_filtered_ad_nhv2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09277296",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    concat_filtered_ad_nhv2,\n",
    "    color=[\n",
    "        \"total_counts\",\n",
    "        \"n_genes_by_counts\",\n",
    "        \"leiden\",\n",
    "    ],\n",
    "    wspace=0.4,\n",
    "    save=\"_v2_Xeniumconcat_withoutharmony.png\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7adce3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    concat_filtered_ad_nhv2,\n",
    "    color=\"batch\",\n",
    "    wspace=0.4,\n",
    "    save=\"_v2_Xeniumconcat_withoutharmony_splitbybatch.png\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffe18078",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save object with UMAP\n",
    "concat_filtered_ad_nhv2.write_h5ad('/path/DataObjects_withUMAP/Xeniumconcat_withoutharmony_umapv2_231210.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5803473",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "squidpy_integration_2",
   "language": "python",
   "name": "squidpy_integration_2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
