{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57ebbf6d",
   "metadata": {},
   "source": [
    "#### This particular notebook details the preprocessing pipeline for our Xenium Dataset 1 (rep 1) data\n",
    "\n",
    "#### Required input files:\n",
    "\n",
    "* Raw Xenium Dataset 1 290 IntReps1and2 data object and filter to only keep Rep1 (availabile via FigShare) OR raw data files for Xenium Dataset 1 290 Rep1 (available via GEO)\n",
    "\n",
    "This notebook starts with the raw GEO files\n",
    "\n",
    "Environment: Please create and activate the conda environment provided in default_env.yaml before running this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd1e7b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "\n",
    "import gzip\n",
    "import anndata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2b37fb9",
   "metadata": {},
   "source": [
    "### Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e76c2a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_10x_h5(\n",
    "    filename=\"/path/cell_feature_matrix.h5\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9beb3b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can unzip cells.csz.gz to obtain cells.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a820725",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    \"/path/cells.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d41a88e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.set_index(adata.obs_names, inplace=True)\n",
    "adata.obs = df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da315d1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm[\"spatial\"] = adata.obs[[\"x_centroid\", \"y_centroid\"]].copy().to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "790eef89",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9458383",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6903d6af",
   "metadata": {},
   "source": [
    "### Calculate Quality Control Metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef5f344f",
   "metadata": {},
   "source": [
    "Calculate the quality control metrics on the anndata.AnnData using scanpy.pp.calculate_qc_metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "919da246",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(adata, percent_top=(10, 20, 50, 150), inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "658e2ad7",
   "metadata": {},
   "source": [
    "The percentage of control probes and control codewords can be calculated from adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d04b4f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cprobes = (\n",
    "    adata.obs[\"control_probe_counts\"].sum() / adata.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "cwords = (\n",
    "    adata.obs[\"control_codeword_counts\"].sum() / adata.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "print(f\"Negative DNA probe count % : {cprobes}\")\n",
    "print(f\"Negative decoding count % : {cwords}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5760b0d0",
   "metadata": {},
   "source": [
    "Next we plot the distribution of total transcripts per cell, unique transcripts per cell, area of segmented cells and the ratio of nuclei area to their cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d57a586a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 4, figsize=(15, 4))\n",
    "\n",
    "axs[0].set_title(\"Total transcripts per cell\")\n",
    "sns.histplot(\n",
    "    adata.obs[\"total_counts\"],\n",
    "    kde=False,\n",
    "    ax=axs[0],\n",
    ")\n",
    "\n",
    "axs[1].set_title(\"Unique transcripts per cell\")\n",
    "sns.histplot(\n",
    "    adata.obs[\"n_genes_by_counts\"],\n",
    "    kde=False,\n",
    "    ax=axs[1],\n",
    ")\n",
    "\n",
    "\n",
    "axs[2].set_title(\"Area of segmented cells\")\n",
    "sns.histplot(\n",
    "    adata.obs[\"cell_area\"],\n",
    "    kde=False,\n",
    "    ax=axs[2],\n",
    ")\n",
    "\n",
    "axs[3].set_title(\"Nucleus ratio\")\n",
    "sns.histplot(\n",
    "    adata.obs[\"nucleus_area\"] / adata.obs[\"cell_area\"],\n",
    "    kde=False,\n",
    "    ax=axs[3],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e21bfa7c",
   "metadata": {},
   "source": [
    "Filter the cells based on the minimum number of counts required using scanpy.pp.filter_cells. Filter the genes based on the minimum number of cells required with scanpy.pp.filter_genes. The parameters for the both were specified based on the plots above. They were set to filter out the cells and genes with minimum counts and minimum cells respectively.\n",
    "\n",
    "Other filter criteria might be cell area, DAPI signal or a minimum of unique transcripts.\n",
    "\n",
    "Squidpy tutorial filtering examples: \n",
    "sc.pp.filter_cells(adata, min_counts=10);\n",
    "sc.pp.filter_genes(adata, min_cells=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d64119c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Our filtering schema\n",
    "# Note: Transcripts that didn't pass 10x Genomics's QC have already been filtered out\n",
    "adata_filtered = adata\n",
    "\n",
    "# filter out cells with <50 counts and <10 genes\n",
    "sc.pp.filter_cells(adata_filtered, min_counts=50)\n",
    "sc.pp.filter_cells(adata_filtered, min_genes=10)\n",
    "\n",
    "# filter out genes that have <1 count and are detected in <10 cells\n",
    "sc.pp.filter_genes(adata_filtered, min_counts=1)\n",
    "sc.pp.filter_genes(adata_filtered, min_cells=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0477b053",
   "metadata": {},
   "source": [
    "Visualize genes with the highest expression levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5566aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highest_expr_genes(adata_filtered, n_top=20, )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2358a5d8",
   "metadata": {},
   "source": [
    "Make a copy of the original raw counts (post-filtering; pre-normalization)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a13e2551",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_filtered.layers['raw_counts'] = adata_filtered.X.copy() # Make a copy\n",
    "adata_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce19e63c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_filtered.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12d567a6",
   "metadata": {},
   "source": [
    "### Save object\n",
    "\n",
    "This is post-filtering and pre-normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22a97412",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_filtered.write_h5ad('/path/DataObjects_withoutUMAP/Xeniumdata_filtered_pre-normalization.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cacfd74",
   "metadata": {},
   "source": [
    "### Continue with analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b268c1de",
   "metadata": {},
   "source": [
    "Normalize counts per cell using scanpy.pp.normalize_total.\n",
    "\n",
    "Logarithmize, do principal component analysis, compute a neighborhood graph of the observations using scanpy.pp.log1p, scanpy.pp.pca and scanpy.pp.neighbors respectively.\n",
    "\n",
    "Use scanpy.tl.umap to embed the neighborhood graph of the data and cluster the cells into subgroups employing scanpy.tl.leiden."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c224f22",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata_filtered, inplace=True)\n",
    "sc.pp.log1p(adata_filtered)\n",
    "# Save log_normalized_counts as a layer\n",
    "adata_filtered.layers['log_normalized_counts']=adata_filtered.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a81450",
   "metadata": {},
   "source": [
    "Calculate and plot the top highly variable genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3bc8616",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note: These min and max values are the default values\n",
    "sc.pp.highly_variable_genes(adata_filtered, min_mean=0.0125, max_mean=3, min_disp=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0d56f88",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highly_variable_genes(adata_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb7ec23f",
   "metadata": {},
   "source": [
    "Regress out unwanted sources or variation and scale data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "258d4502",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here, we're going to regress out total_counts and n_genes_by_counts\n",
    "\n",
    "sc.pp.regress_out(adata_filtered, [\"total_counts\",\"n_genes_by_counts\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ae2b790",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.scale(adata_filtered, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4de911e3",
   "metadata": {},
   "source": [
    "Run PCA and plot PCA variance ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f64ebf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.pca(adata_filtered, svd_solver='arpack')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fc95196",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca(adata_filtered, color=['EPCAM','COL3A1','CD69'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c2ef334",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca_variance_ratio(adata_filtered, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ea32d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca_variance_ratio(adata_filtered, n_pcs = 50, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dca4c63e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48f24143",
   "metadata": {},
   "source": [
    "### Compute neighbor graph and plot UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edbf66c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.figdir = '/path/UMAP_pngs/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae30692f",
   "metadata": {},
   "source": [
    "#### Parameter descriptions:\n",
    "n_neighbors\n",
    "* A value between 2 and 100, representing the number of neighboring data points used for manifld approximation. Larger values give a manifold with a more global view of the dataset, while smaller values preserve more of the local structures.\n",
    "* Default value is 15\n",
    "\n",
    "n_pcs\n",
    "* Use this many PCs\n",
    "* Default value is None\n",
    "\n",
    "min_dist\n",
    "* The minimum distance between two points in the UMAP embedding.\n",
    "* Default value is 0.05\n",
    "\n",
    "spread\n",
    "* A scaling factor for distance between embedded points.\n",
    "* Default value is 1.0\n",
    "\n",
    "Helpful resource: https://smorabit.github.io/blog/2020/umap/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8ef9b7f",
   "metadata": {},
   "source": [
    "#### Different versions -- Playing around with different n_pcs and spread values\n",
    "\n",
    "#### v4: \n",
    "n_neighbors = 10; \n",
    "n_pcs = 30;\n",
    "min_dist = 0.02;\n",
    "spread = 1.75"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e5ce011",
   "metadata": {},
   "source": [
    "#### Testing v4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d646ee1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_filtered_v4 = adata_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac352213",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata_filtered_v4, n_neighbors=10, n_pcs=30)\n",
    "sc.tl.umap(adata_filtered_v4, min_dist=0.02, spread=1.75)\n",
    "sc.tl.leiden(adata_filtered_v4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66f283d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata_filtered_v4,\n",
    "    color=[\n",
    "        \"total_counts\",\n",
    "        \"n_genes_by_counts\",\n",
    "        \"leiden\",\n",
    "    ],\n",
    "    wspace=0.4,\n",
    "    save = '_v4_Xeniumdata.png',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d7bceaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save object with UMAP\n",
    "adata_filtered_v4.write_h5ad('/path/DataObjects_withUMAP/Xeniumdata_umapv4.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf7cf590",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "squidpy_plot",
   "language": "python",
   "name": "squidpy_plot"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
