{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c51ac90c",
   "metadata": {},
   "source": [
    "#### This particular notebook includes a pipeline for generating filtered transcript files (to be used for QC comparisons) and saving a gene list csv file from a Xenium run. This example uses Xenium Dataset 3 slide 1.\n",
    "\n",
    "#### Required input files:\n",
    "* Raw transcript csv file\n",
    "* Filtered cell-based data object\n",
    "\n",
    "Note: If you have a raw transcripts.parquet file instead of a .csv.gz, you can use the included ConvertXeniumRawTranscriptsParquetToCsv.sh script to generate a CSV version.\n",
    "\n",
    "Environment: Please create and activate the conda environment provided in default_env.yaml before running this notebook\n",
    "\n",
    "#### Caution: This notebook is highly memory-intensive and may crash during processing. It is recommended to save intermediate outputs at key steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd1e7b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from adjustText import adjust_text\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "\n",
    "import gzip\n",
    "import anndata\n",
    "\n",
    "import sys\n",
    "\n",
    "from matplotlib.ticker import FuncFormatter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e2d2a10",
   "metadata": {},
   "source": [
    "### Filtering Syntax Key\n",
    "\n",
    "transcripts_raw = No filters applied (just automatic technology filters)\n",
    "\n",
    "transcripts_initialfilter = Respective technology filters have been applied, as well as our additional QC filtering\n",
    "\n",
    "transcripts_r = False positive and low quality transcripts have been removed, as well as the respective technology filters and our QC filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2b37fb9",
   "metadata": {},
   "source": [
    "# Slide1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c7dce95",
   "metadata": {},
   "source": [
    "### Xenium Filtering Note\n",
    "\n",
    "10x Genomics performs their own filtering to remove low quality transcripts (Q-score < 20) before outputting cell-associated data files (cell feature matrix) for analysis. These low quality transcripts are not removed from the transcript file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63f9f05a",
   "metadata": {},
   "source": [
    "# Data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66d0d66f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load in cell-based filtered data object (post-filtering, pre-normalization)\n",
    "\n",
    "Int_XeniumICI480_Data = sc.read_h5ad('/path/DataObjects_withoutUMAP/XeniumICI480_concat_filtered_pre-normalization_240603.h5ad')\n",
    "\n",
    "# If applicable, filter based on batch\n",
    "Slide1_XeniumICI480_Data = Int_XeniumICI480_Data[Int_XeniumICI480_Data.obs['slide'] == 'slide_1'].copy()\n",
    "\n",
    "display(Slide1_XeniumICI480_Data.obs)\n",
    "display(Slide1_XeniumICI480_Data.obs['slide'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a77d8e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate min, mean, and max values for Slide1_XeniumICI480_Data\n",
    "\n",
    "total_counts_minvalue = Slide1_XeniumICI480_Data.obs[\"total_counts\"].min()\n",
    "total_counts_mean = Slide1_XeniumICI480_Data.obs[\"total_counts\"].mean()\n",
    "total_counts_maxvalue = Slide1_XeniumICI480_Data.obs[\"total_counts\"].max()\n",
    "\n",
    "n_genes_by_counts_minvalue = Slide1_XeniumICI480_Data.obs[\"n_genes_by_counts\"].min()\n",
    "n_genes_by_counts_mean = Slide1_XeniumICI480_Data.obs[\"n_genes_by_counts\"].mean()\n",
    "n_genes_by_counts_maxvalue = Slide1_XeniumICI480_Data.obs[\"n_genes_by_counts\"].max()\n",
    "\n",
    "cell_area_um2_minvalue = Slide1_XeniumICI480_Data.obs[\"cell_area\"].min()\n",
    "cell_area_um2_mean = Slide1_XeniumICI480_Data.obs[\"cell_area\"].mean()\n",
    "cell_area_um2_maxvalue = Slide1_XeniumICI480_Data.obs[\"cell_area\"].max()\n",
    "\n",
    "nucleus_area_um2_minvalue = Slide1_XeniumICI480_Data.obs[\"nucleus_area\"].min()\n",
    "nucleus_area_um2_mean = Slide1_XeniumICI480_Data.obs[\"nucleus_area\"].mean()\n",
    "nucleus_area_um2_maxvalue = Slide1_XeniumICI480_Data.obs[\"nucleus_area\"].max()\n",
    "\n",
    "print(\"Slide1_XeniumICI480_Data\\n\")\n",
    "\n",
    "print(f\"Min value of total_counts : {total_counts_minvalue}\")\n",
    "print(f\"Mean of total_counts : {total_counts_mean}\")\n",
    "print(f\"Max value of total_counts : {total_counts_maxvalue}\\n\")\n",
    "\n",
    "print(f\"Min value of n_genes_by_counts : {n_genes_by_counts_minvalue}\")\n",
    "print(f\"Mean of n_genes_by_counts : {n_genes_by_counts_mean}\")\n",
    "print(f\"Max value of n_genes_by_counts : {n_genes_by_counts_maxvalue}\\n\")\n",
    "\n",
    "print(f\"Min value of cell area (um2) : {cell_area_um2_minvalue}\")\n",
    "print(f\"Mean of cell area (um2) : {cell_area_um2_mean}\")\n",
    "print(f\"Max value of cell area (um2) : {cell_area_um2_maxvalue}\\n\")\n",
    "\n",
    "print(f\"Min value of nucleus area (um2) : {nucleus_area_um2_minvalue}\")\n",
    "print(f\"Mean of nucleus area (um2) : {nucleus_area_um2_mean}\")\n",
    "print(f\"Max value of nucleus area (um2) : {nucleus_area_um2_maxvalue}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c393cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate transcript-associated metrics\n",
    "\n",
    "# Note: transcript_counts and total_counts should be the same (besides a minimal rounding error)\n",
    "\n",
    "trc = Slide1_XeniumICI480_Data.obs[\"transcript_counts\"].mean()\n",
    "toc = Slide1_XeniumICI480_Data.obs[\"total_counts\"].mean()\n",
    "\n",
    "n_genes_by_counts_mean = Slide1_XeniumICI480_Data.obs[\"n_genes_by_counts\"].mean()\n",
    "\n",
    "print(\"Mean number of transcripts per cell for Slide1_XeniumICI480_Data\")\n",
    "print(\"\")\n",
    "print(f\"Mean transcript_counts value of Slide1_XeniumICI480_Data : {trc}\")\n",
    "print(f\"Mean total_counts value of Slide1_XeniumICI480_Data : {toc}\\n\")\n",
    "\n",
    "print(\"Mean number of unique genes per cell\")\n",
    "print(\"\")\n",
    "print(f\"Mean n_genes_by_counts of Slide1_XeniumICI480_Data : {n_genes_by_counts_mean}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6229ed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Caluclate total cell area\n",
    "\n",
    "cell_area_um2_sum = Slide1_XeniumICI480_Data.obs['cell_area'].sum()\n",
    "\n",
    "print(f\"Cell area sum (um2) : {cell_area_um2_sum}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39614514",
   "metadata": {},
   "source": [
    "The negative control values included in Xenium cell-based metadata:\n",
    "* Control probe counts (probes that exist in the panel but target non-biological sequences)\n",
    "* Control codeword counts (codewords that don't have any probes matching that code)\n",
    "* Unassigned codeword counts (there is no probe in the panel that will generate the codeword)\n",
    "* Deprecated codeword counts (assigned to codewords that are not used in the Xenium Onboard Analysis pipeline)\n",
    "\n",
    "The percentage of these negative control values within the dataset can be calculated from Slide1_XeniumICI480_Data.obs\n",
    "\n",
    "\n",
    "#### Reminder:\n",
    "\n",
    "If you ran sc.pp.calculate_qc_metrics() on your Xenium cell-based data, then the total_counts metadata column is adjusted to only include the total number of transcripts per cell, which only includes gene features.\n",
    "\n",
    "Originally, before running sc.pp.calculate_qc_metrics() (or if you didn't run it), then the total_counts metadata column includes a sum total of the gene expression features, control_probe_counts, control_codeword_counts, and unassigned_codeword_counts.\n",
    "\n",
    "This notebook will use the total_counts value that has been adjusted by sc.pp.caluclate_qc_metrics() to only include the sum total of the gene expression features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bebc875f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate negative control-associated metrics\n",
    "\n",
    "# Percents\n",
    "\n",
    "cprobes_p = (\n",
    "    Slide1_XeniumICI480_Data.obs[\"control_probe_counts\"].sum() / Slide1_XeniumICI480_Data.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "cwords_p = (\n",
    "    Slide1_XeniumICI480_Data.obs[\"control_codeword_counts\"].sum() / Slide1_XeniumICI480_Data.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "\n",
    "# False positives\n",
    "uwords_p = (\n",
    "    Slide1_XeniumICI480_Data.obs[\"unassigned_codeword_counts\"].sum() / Slide1_XeniumICI480_Data.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "\n",
    "dwords_p = (\n",
    "    Slide1_XeniumICI480_Data.obs[\"deprecated_codeword_counts\"].sum() / Slide1_XeniumICI480_Data.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "\n",
    "print(f\"Percents\")\n",
    "print(f\"Negative DNA probe count % : {cprobes_p}\")\n",
    "print(f\"Negative decoding count % : {cwords_p}\")\n",
    "print(f\"Unassigned codeword count % : {uwords_p}\")\n",
    "print(f\"Deprecated codeword count % : {dwords_p}\\n\")\n",
    "\n",
    "# Sums\n",
    "\n",
    "cprobes_s = Slide1_XeniumICI480_Data.obs[\"control_probe_counts\"].sum()\n",
    "\n",
    "cwords_s = Slide1_XeniumICI480_Data.obs[\"control_codeword_counts\"].sum()\n",
    "\n",
    "uwords_s = Slide1_XeniumICI480_Data.obs[\"unassigned_codeword_counts\"].sum()\n",
    "\n",
    "dwords_s = Slide1_XeniumICI480_Data.obs[\"deprecated_codeword_counts\"].sum()\n",
    "\n",
    "print(f\"Sums\")\n",
    "print(f\"Negative DNA probe count # : {cprobes_s}\")\n",
    "print(f\"Negative decoding count # : {cwords_s}\")\n",
    "print(f\"Unassigned codeword count # : {uwords_s}\")\n",
    "print(f\"Deprecated codeword count # : {dwords_s}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df37e572",
   "metadata": {},
   "outputs": [],
   "source": [
    "## A metric that we've liked using to assess run quality\n",
    "\n",
    "Slide1_XeniumICI480_Data.obs[\"control_probe_counts\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dae27da0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 10x Genomics's recommended false positive metric\n",
    "# Calculated by: Mean negative probe count per cell / number of control probes / number of cells\n",
    "\n",
    "# Example from our XeniumICI480 dataset: (0.010 / 20 / 539650)\n",
    "\n",
    "# INSERT HERE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba89c3aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Print gene list\n",
    "# Note: Gene list (.var['gene_ids']) got lost when concatenating the data, but it's still stored as .var_names so we can pull it from there\n",
    "\n",
    "Slide1_XeniumICI480_Data.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a104176",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Keep when applicable (new panel)\n",
    "\n",
    "# Convert to a table\n",
    "XeniumICI480_GeneList = pd.DataFrame(Slide1_XeniumICI480_Data.var_names)\n",
    "\n",
    "# Name gene column\n",
    "XeniumICI480_GeneList.rename(columns={0: \"unique_genes\"}, inplace=True)\n",
    "\n",
    "# Print\n",
    "XeniumICI480_GeneList\n",
    "\n",
    "# Ensure that this count matches your expected number of genes. Otherwise, genes may have been removed during cell or transcript filtering steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "043c35fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Save if desired\n",
    "\n",
    "#XeniumICI480_GeneList.to_csv('/path/XeniumICI_Custom480Panel_GeneList.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f58335f1",
   "metadata": {},
   "source": [
    "# Transcript File\n",
    "\n",
    "Read in Xenium transcript file and visualize it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "799b5886",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in raw transcript file and visualize\n",
    "\n",
    "# Xenium (Note: Unzip if necessary. In this case, the file had already been unzipped and saved by using our parquet env)\n",
    "Slide1_XeniumICI480_transcripts_raw = pd.read_csv('/path/SlideID_0063975_transcripts.csv')\n",
    "\n",
    "Slide1_XeniumICI480_transcripts_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "340a8559",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quantify number of fov's\n",
    "\n",
    "n_fovs = Slide1_XeniumICI480_transcripts_raw['fov_name'].nunique()\n",
    "\n",
    "print(f\"Number of unique fov's : {n_fovs}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b790a4f8",
   "metadata": {},
   "source": [
    "### Merge Slide1_XeniumICI480_Data (from above) with Xenium_transcripts_raw. This way, we're working with Xenium_transcripts_initialfilter (transcripts within cells that passed 10x's QC filter and our filtering)\n",
    "\n",
    "This is also important for Xenium runs with orientation cores. It ensures that we remove cells from the orientation cores (which were removed from the cell-focused object.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a223c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# View metadata\n",
    "Slide1_XeniumICI480_Data.obs['cell_id']\n",
    "\n",
    "# Note: These are all of the cells that passed 10x QC and our QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4701d933",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Format for merge\n",
    "\n",
    "# Need to remove the index name because there's also a column with the same name (cell_id)\n",
    "Slide1_XeniumICI480_Data.obs.index.name = None\n",
    "\n",
    "# Rename 'cell_id' to 'cell_id_wSlideName'\n",
    "Slide1_XeniumICI480_Data.obs.rename(columns={\"cell_id\": \"cell_id_wSlideName\"}, inplace=True)\n",
    "\n",
    "# Duplicate the cell_id_wSlideName column and strip the prefix (so that the cell ids match the transcript ids)\n",
    "\n",
    "Slide1_XeniumICI480_Data.obs[\"cell_id\"] = (\n",
    "    Slide1_XeniumICI480_Data.obs[\"cell_id_wSlideName\"].str.replace(\"^Slide1-\", \"\", regex=True)\n",
    ")\n",
    "\n",
    "# View\n",
    "display(Slide1_XeniumICI480_Data.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b596625d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge the data frames based on \"cell_id\"\n",
    "Slide1_XeniumICI480_transcripts_initialfilter = pd.merge(Slide1_XeniumICI480_transcripts_raw, Slide1_XeniumICI480_Data.obs[['cell_id']], on='cell_id', how='inner')\n",
    "\n",
    "# View\n",
    "Slide1_XeniumICI480_transcripts_initialfilter\n",
    "\n",
    "\n",
    "# Result: End up with 134,672,733 rows (transcripts) instead of 170,805,937\n",
    "\n",
    "\n",
    "## Consider saving Slide1_XeniumICI480_transcripts_initialfilter here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "880e3059",
   "metadata": {},
   "source": [
    "### Done - Continue with Xenium_transcripts_initialfilter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d012839b",
   "metadata": {},
   "source": [
    "### Create new objects - Remove false positives / low quality / other undesirable values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70a46004",
   "metadata": {},
   "outputs": [],
   "source": [
    "## First, we actually want to create an object with just the negative probe transcripts (just to compare later on to the gene probe mean counts)\n",
    "\n",
    "# Make a copy\n",
    "Slide1_XeniumICI480_negprobe_transcripts = Slide1_XeniumICI480_transcripts_initialfilter.copy()\n",
    "\n",
    "# Keep only the rows where 'feature_name' contains 'NegControlProbe'\n",
    "Slide1_XeniumICI480_negprobe_transcripts = Slide1_XeniumICI480_negprobe_transcripts[Slide1_XeniumICI480_negprobe_transcripts['feature_name'].str.contains('NegControlProbe', case=False, na=False)]\n",
    "\n",
    "# View\n",
    "Slide1_XeniumICI480_negprobe_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a522a89",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Make new df\n",
    "Slide1_XeniumICI480_negprobe_transcripts2 = Slide1_XeniumICI480_negprobe_transcripts.copy()\n",
    "\n",
    "# Remove rows where 'cell_id' matches 'UNASSIGNED'\n",
    "Slide1_XeniumICI480_negprobe_transcripts2 = Slide1_XeniumICI480_negprobe_transcripts2[Slide1_XeniumICI480_negprobe_transcripts2['cell_id'] != 'UNASSIGNED']\n",
    "\n",
    "# Remove transcripts that didn't pass QC\n",
    "Slide1_XeniumICI480_negprobe_transcripts2 = Slide1_XeniumICI480_negprobe_transcripts2[Slide1_XeniumICI480_negprobe_transcripts2['qv'] >= 20.0]\n",
    "\n",
    "# Group by cell\n",
    "Slide1_XeniumICI480_negprobe_transcripts2 = Slide1_XeniumICI480_negprobe_transcripts2.groupby('cell_id')['transcript_id'].nunique().reset_index()\n",
    "\n",
    "# Rename columns for clarity\n",
    "Slide1_XeniumICI480_negprobe_transcripts2.columns = ['cell_id', 'negprobe_count']\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "print(Slide1_XeniumICI480_negprobe_transcripts2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b7a1876",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate negprobe_count sum\n",
    "\n",
    "Slide1_XeniumICI480_negprobe_transcripts2['negprobe_count'].sum()\n",
    "\n",
    "# Note: This matches what we found in the metadata for the negative control probe counts (scroll up)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "619e21cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Looking for other controls\n",
    "\n",
    "Xenium_control_transcripts = Slide1_XeniumICI480_transcripts_initialfilter.copy()\n",
    "\n",
    "# Filter based on feature_name value including an '_'\n",
    "filtered_featurenames_controls = Xenium_control_transcripts['feature_name'][Xenium_control_transcripts['feature_name'].str.contains('_')]\n",
    "\n",
    "# Get unique values\n",
    "unique_featurenames_controls = filtered_featurenames_controls.unique()\n",
    "\n",
    "# Alphabetize\n",
    "sorted_unique_featurenames_controls = sorted(unique_featurenames_controls)\n",
    "\n",
    "# Print\n",
    "for value in sorted_unique_featurenames_controls:\n",
    "    print(value)\n",
    "    \n",
    "    \n",
    "# Note: There are NegControlCodeword and NegControlProbe -- No instances of DeprecatedCodeword, Intergenic_Region, or UnassignedCodeword (as seen in our Xenium datasets)\n",
    "# The controls per Xenium dataset may vary based on the chemistry version and 10x updates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fecedbda",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Looking for other controls, continued\n",
    "\n",
    "featurenames_blanks = Slide1_XeniumICI480_transcripts_initialfilter.copy()\n",
    "\n",
    "filtered_featurenames_blanks = featurenames_blanks[(featurenames_blanks['feature_name'].str.contains('BLANK', case=False, na=False))]\n",
    "\n",
    "# print\n",
    "filtered_featurenames_blanks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3a5c4a3",
   "metadata": {},
   "source": [
    "Okay, moving forward.\n",
    "\n",
    "We will re-name this initialfilter object to r, to denote filtering removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf3aab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove false positives -- Modify this as needed\n",
    "\n",
    "Slide1_XeniumICI480_transcripts_r = Slide1_XeniumICI480_transcripts_initialfilter.copy()\n",
    "\n",
    "# Remove rows where 'cell_id' matches 'UNASSIGNED'\n",
    "Slide1_XeniumICI480_transcripts_r = Slide1_XeniumICI480_transcripts_r[Slide1_XeniumICI480_transcripts_r['cell_id'] != 'UNASSIGNED']\n",
    "\n",
    "# Remove rows where 'feature_name' contains '_' (All of the controls)\n",
    "Slide1_XeniumICI480_transcripts_r = Slide1_XeniumICI480_transcripts_r[~(Slide1_XeniumICI480_transcripts_r['feature_name'].str.contains('_', case=False, na=False))]\n",
    "\n",
    "# Remove rows where 'feature_name' contains 'BLANK'\n",
    "Slide1_XeniumICI480_transcripts_r = Slide1_XeniumICI480_transcripts_r[~(Slide1_XeniumICI480_transcripts_r['feature_name'].str.contains('BLANK', case=False, na=False))]\n",
    "\n",
    "# Remove transcripts that didn't pass QC\n",
    "Slide1_XeniumICI480_transcripts_r = Slide1_XeniumICI480_transcripts_r[Slide1_XeniumICI480_transcripts_r['qv'] >= 20.0]\n",
    "\n",
    "# View\n",
    "Slide1_XeniumICI480_transcripts_r\n",
    "\n",
    "\n",
    "## 117,335,771 transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89bd8929",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Print the number of unique genes\n",
    "\n",
    "Slide1_XeniumICI480_transcripts_r['feature_name'].nunique()\n",
    "\n",
    "# Ensure that this value is the expected number of genes in the panel (480) -- YES\n",
    "# Meaning that all of the controls have been removed. And that all of the genes survived the cell-specific filtering\n",
    "# If a gene was filtered out during the cell-specific filtering, we need to figure out which gene that is\n",
    "# Run the code below if that's the case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "803ab1c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save Slide1_XeniumICI480_transcripts_r\n",
    "\n",
    "#Slide1_XeniumICI480_transcripts_r.to_csv('/path/Slide1_XeniumICI480_transcripts_r.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f25a4be4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42238e3a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "squidpy_plot",
   "language": "python",
   "name": "squidpy_plot"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
