{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "77250e68",
   "metadata": {},
   "source": [
    "#### This particular notebook includes QC metric calculations and comparisons between: \n",
    "* Dataset 1: Xenium (rep 1 only) vs CosMx\n",
    "* Dataset 1: Xenium rep 1 vs rep 2 (occasionally referred to as 'Sept' and 'Nov' due to their run dates)\n",
    "\n",
    "It can be adapted to compare any 2 datasets or reps/slides within a dataset. \n",
    "\n",
    "\n",
    "#### Required input files:\n",
    "* Overlapping gene list between datasets (if applicable)\n",
    "* Filtered transcript file (for each dataset/rep/slide)\n",
    "* Filtered cell-based data object (processed or unprocessed) (for each dataset)\n",
    "\n",
    "Note: r denotes filtered transcript file (filtering removal)\n",
    "\n",
    "Environment: Please create and activate the conda environment provided in default_env.yaml before running this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f381d7a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load in packages\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.transforms as mtransforms\n",
    "import seaborn as sns\n",
    "\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "\n",
    "import gzip\n",
    "import anndata\n",
    "\n",
    "import os\n",
    "\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import pearsonr\n",
    "from scipy.stats import kruskal\n",
    "\n",
    "from adjustText import adjust_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "159d9461",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set output directory\n",
    "\n",
    "sc.settings.figdir = '/path/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "989e53a9",
   "metadata": {},
   "source": [
    "## Transcript-Related Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9af4eb2d",
   "metadata": {},
   "source": [
    "#### Generate data objects for plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "964e36a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in overlapping genes list between datasets\n",
    "overlappinggenes_df = pd.read_csv('/path/CosMx1000Xenium290_overlappinggenelist_231112.csv')\n",
    "\n",
    "# Convert the merged DataFrame to a list\n",
    "overlappinggenes_list = overlappinggenes_df['unique_genes'].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46c2da92",
   "metadata": {},
   "source": [
    "Rep1_Xenium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3809e738",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in data\n",
    "\n",
    "Xenium_transcripts_r = pd.read_csv('/path/Xenium_transcripts_r.csv')\n",
    "\n",
    "Xenium_transcripts_r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb757a6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantifying cellular location classification for transcripts\n",
    "\n",
    "# Xenium_transcripts_r. Percentages of whole\n",
    "\n",
    "# 46486849 total transcripts\n",
    "\n",
    "Xenium_r_overlaps_nucleus_0no = (((Xenium_transcripts_r['overlaps_nucleus'] == 0).sum())/46486849) * 100\n",
    "Xenium_r_overlaps_nucleus_1yes = (((Xenium_transcripts_r['overlaps_nucleus'] == 1).sum())/46486849) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f8c3265",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantify number of transcripts per cell\n",
    "\n",
    "## To answer the question: How many transcripts does each cell have?\n",
    "\n",
    "# Make new df\n",
    "Xenium_transcript_counts_per_cell_r = Xenium_transcripts_r.groupby('cell_id')['transcript_id'].nunique().reset_index()\n",
    "\n",
    "# Rename columns for clarity\n",
    "Xenium_transcript_counts_per_cell_r.columns = ['cell_id', 'transcript_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b70fe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of transcripts per cell -- All genes\")\n",
    "print(f\"Xenium, Mean: {Xenium_transcript_counts_per_cell_r['transcript_count'].mean()}\")\n",
    "print(f\"Xenium, Median: {Xenium_transcript_counts_per_cell_r['transcript_count'].median()}\")\n",
    "print(f\"Xenium, Q1: {Xenium_transcript_counts_per_cell_r['transcript_count'].quantile(0.25)}\")\n",
    "print(f\"Xenium, Q3: {Xenium_transcript_counts_per_cell_r['transcript_count'].quantile(0.75)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2781b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## What about for only the overlapping genes?\n",
    "\n",
    "## To answer the question: How many transcripts (of overlapping genes) does each cell have?\n",
    "\n",
    "\n",
    "### Part 1: Make overlapping genes df\n",
    "\n",
    "## Filter df to only keep rows with overlapping genes\n",
    "\n",
    "# Create a boolean mask to filter rows based on the selected genes\n",
    "Xenium_mask_r = Xenium_transcripts_r['feature_name'].isin(overlappinggenes_list)\n",
    "\n",
    "# Apply the mask to filter the DataFrame\n",
    "Xenium_transcripts_r_overlappinggenes = Xenium_transcripts_r[Xenium_mask_r]\n",
    "\n",
    "### Part 2: Quantify transcripts for overlapping genes df\n",
    "\n",
    "# Make new df\n",
    "Xenium_transcript_counts_per_cell_overlappinggenes_r = Xenium_transcripts_r_overlappinggenes.groupby('cell_id')['transcript_id'].nunique().reset_index()\n",
    "\n",
    "# Rename columns for clarity\n",
    "Xenium_transcript_counts_per_cell_overlappinggenes_r.columns = ['cell_id', 'transcript_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e46fb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of transcripts per cell -- Overlapping genes\")\n",
    "print(f\"Xenium, Mean: {Xenium_transcript_counts_per_cell_overlappinggenes_r['transcript_count'].mean()}\")\n",
    "print(f\"Xenium, Median: {Xenium_transcript_counts_per_cell_overlappinggenes_r['transcript_count'].median()}\")\n",
    "print(f\"Xenium, Q1: {Xenium_transcript_counts_per_cell_overlappinggenes_r['transcript_count'].quantile(0.25)}\")\n",
    "print(f\"Xenium, Q3: {Xenium_transcript_counts_per_cell_overlappinggenes_r['transcript_count'].quantile(0.75)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb48d64a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Table with cell id and feature name values\n",
    "\n",
    "# Make df with just cell_id and feature_name\n",
    "Xenium_transcripts_r_featurespercell = Xenium_transcripts_r[['cell_id', 'feature_name']]\n",
    "\n",
    "## Print unique features per cell\n",
    "\n",
    "## To answer the question: What unique features does each cell express?\n",
    "\n",
    "# Drop duplicate rows with same cell_id and feature_name combo\n",
    "Xenium_transcripts_r_uniquefeaturespercell = Xenium_transcripts_r_featurespercell.drop_duplicates(subset=['cell_id', 'feature_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec0fe913",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantify the number of unique feature_name values per cell\n",
    "\n",
    "## To answer the question: How many unique features does each cell express\n",
    "\n",
    "# Make new df\n",
    "Xenium_transcripts_r_uniquefeaturespercell_count = Xenium_transcripts_r_uniquefeaturespercell.groupby('cell_id')['feature_name'].nunique().reset_index()\n",
    "\n",
    "# Rename the new column for clarity\n",
    "Xenium_transcripts_r_uniquefeaturespercell_count.columns = ['cell_id', 'unique_features']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8a09f15",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of unique features per cell -- All genes\")\n",
    "print(f\"Xenium, Mean: {Xenium_transcripts_r_uniquefeaturespercell_count['unique_features'].mean()}\")\n",
    "print(f\"Xenium, Median: {Xenium_transcripts_r_uniquefeaturespercell_count['unique_features'].median()}\")\n",
    "print(f\"Xenium, Q1: {Xenium_transcripts_r_uniquefeaturespercell_count['unique_features'].quantile(0.25)}\")\n",
    "print(f\"Xenium, Q3: {Xenium_transcripts_r_uniquefeaturespercell_count['unique_features'].quantile(0.75)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8fd45a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## What about for only overlapping genes?\n",
    "\n",
    "## To answer the question: How many unique features (of overlapping genes) does each cell have?\n",
    "\n",
    "\n",
    "### Part 1: Get rid of duplicate feature rows\n",
    "\n",
    "# Drop duplicate rows with same cell_id and feature_name combo\n",
    "Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes = Xenium_transcripts_r_overlappinggenes.drop_duplicates(subset=['cell_id', 'feature_name'])\n",
    "\n",
    "### Part 2: Quantify unique features for overlapping genes df\n",
    "\n",
    "# Make new df\n",
    "Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes_count = Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes.groupby('cell_id')['feature_name'].nunique().reset_index()\n",
    "\n",
    "# Rename the new column for clarity\n",
    "Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes_count.columns = ['cell_id', 'unique_features']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bc3cf0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of unique features per cell -- Overlapping genes\")\n",
    "print(f\"Xenium, Mean: {Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].mean()}\")\n",
    "print(f\"Xenium, Median: {Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].median()}\")\n",
    "print(f\"Xenium, Q1: {Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].quantile(0.25)}\")\n",
    "print(f\"Xenium, Q3: {Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].quantile(0.75)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6827eca",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantify the frequency / count that each feature appears in each cell (gene probe expression level)\n",
    "\n",
    "## To answer the question: What is the expression level of each feature in each cell? (Per gene, per cell)\n",
    "\n",
    "# Make a copy\n",
    "Xenium_transcripts_r_quantityofeachfeaturepercell = Xenium_transcripts_r_featurespercell.copy()\n",
    "\n",
    "# Quantify feature appearances\n",
    "Xenium_grouped_counts = Xenium_transcripts_r_quantityofeachfeaturepercell.groupby(['cell_id', 'feature_name'])['feature_name'].transform('count')\n",
    "Xenium_transcripts_r_quantityofeachfeaturepercell.loc[:, 'feature_quantity'] = Xenium_grouped_counts\n",
    "\n",
    "# Drop duplicate rows to keep only unique combinations\n",
    "Xenium_transcripts_r_quantityofeachfeaturepercell = Xenium_transcripts_r_quantityofeachfeaturepercell.drop_duplicates(subset=['cell_id', 'feature_name'])\n",
    "\n",
    "# Resetting index for clarity\n",
    "Xenium_transcripts_r_quantityofeachfeaturepercell.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d42e0181",
   "metadata": {},
   "outputs": [],
   "source": [
    "## What about for only the overlapping genes?\n",
    "\n",
    "## To answer the question: What is the expression level of each overlapping feature in each cell? (Per gene, per cell)\n",
    "\n",
    "## Filter df to only keep rows with overlapping genes\n",
    "\n",
    "# Create a boolean mask to filter rows based on the selected genes\n",
    "Xenium_mask_r = Xenium_transcripts_r_quantityofeachfeaturepercell['feature_name'].isin(overlappinggenes_list)\n",
    "\n",
    "# Apply the mask to filter the DataFrame\n",
    "Xenium_transcripts_r_quantityofeachfeaturepercell_overlappinggenes = Xenium_transcripts_r_quantityofeachfeaturepercell[Xenium_mask_r]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6125e9",
   "metadata": {},
   "source": [
    "Rep2_Xenium (sometimes referred to as 'Nov' for 'November run')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dcf9ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in data\n",
    "\n",
    "Rep2_Xenium_transcripts_r = pd.read_csv('/path/Nov_Xenium_transcripts_r.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c76b809",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Table with cell id and feature name values\n",
    "\n",
    "# Make df with just cell_id and feature_name\n",
    "Rep2_Xenium_transcripts_r_featurespercell = Rep2_Xenium_transcripts_r[['cell_id', 'feature_name']]\n",
    "\n",
    "## Print unique features per cell\n",
    "\n",
    "## To answer the question: What unique features does each cell express?\n",
    "\n",
    "# Drop duplicate rows with same cell_id and feature_name combo\n",
    "Rep2_Xenium_transcripts_r_uniquefeaturespercell = Rep2_Xenium_transcripts_r_featurespercell.drop_duplicates(subset=['cell_id', 'feature_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c753136c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantify the frequency / count that each feature appears in each cell (gene probe expression level)\n",
    "\n",
    "## To answer the question: What is the expression level of each feature in each cell? (Per gene, per cell)\n",
    "\n",
    "# Make a copy\n",
    "Rep2_Xenium_transcripts_r_quantityofeachfeaturepercell = Rep2_Xenium_transcripts_r_featurespercell.copy()\n",
    "\n",
    "# Quantify feature appearances\n",
    "Rep2_Xenium_grouped_counts = Rep2_Xenium_transcripts_r_quantityofeachfeaturepercell.groupby(['cell_id', 'feature_name'])['feature_name'].transform('count')\n",
    "Rep2_Xenium_transcripts_r_quantityofeachfeaturepercell.loc[:, 'feature_quantity'] = Rep2_Xenium_grouped_counts\n",
    "\n",
    "# Drop duplicate rows to keep only unique combinations\n",
    "Rep2_Xenium_transcripts_r_quantityofeachfeaturepercell = Rep2_Xenium_transcripts_r_quantityofeachfeaturepercell.drop_duplicates(subset=['cell_id', 'feature_name'])\n",
    "\n",
    "# Resetting index for clarity\n",
    "Rep2_Xenium_transcripts_r_quantityofeachfeaturepercell.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d41655",
   "metadata": {},
   "source": [
    "CosMx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebbfd14b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in data\n",
    "\n",
    "CosMx_transcripts_r = pd.read_csv('/CosMx_transcripts_r.csv')\n",
    "\n",
    "CosMx_transcripts_r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3cd9da",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantifying cellular location classification for transcripts\n",
    "\n",
    "# CosMx_transcripts_r. Percentages of whole\n",
    "\n",
    "# 27695490 total transcripts\n",
    "\n",
    "CosMx_r_pct_nuc = (((CosMx_transcripts_r['CellComp'] == 'Nuclear').sum())/27695490) * 100\n",
    "CosMx_r_pct_mem = (((CosMx_transcripts_r['CellComp'] == 'Membrane').sum())/27695490) * 100\n",
    "CosMx_r_pct_cyt = (((CosMx_transcripts_r['CellComp'] == 'Cytoplasm').sum())/27695490) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bae10ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantify number of transcripts per cell\n",
    "\n",
    "## To answer the question: How many transcripts does each cell have?\n",
    "\n",
    "# Make new df\n",
    "CosMx_transcript_counts_per_cell_r = CosMx_transcripts_r.groupby('cell_id')['transcript_id'].nunique().reset_index()\n",
    "\n",
    "# Rename columns for clarity\n",
    "CosMx_transcript_counts_per_cell_r.columns = ['cell_id', 'transcript_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85d948e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of transcripts per cell -- All genes\")\n",
    "print(f\"CosMx, Mean: {CosMx_transcript_counts_per_cell_r['transcript_count'].mean()}\")\n",
    "print(f\"CosMx, Median: {CosMx_transcript_counts_per_cell_r['transcript_count'].median()}\")\n",
    "print(f\"CosMx, Q1: {CosMx_transcript_counts_per_cell_r['transcript_count'].quantile(0.25)}\")\n",
    "print(f\"CosMx, Q3: {CosMx_transcript_counts_per_cell_r['transcript_count'].quantile(0.75)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "354b06d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## What about for only the overlapping genes?\n",
    "\n",
    "## To answer the question: How many transcripts (of overlapping genes) does each cell have?\n",
    "\n",
    "\n",
    "### Part 1: Make overlapping genes df\n",
    "\n",
    "## Filter df to only keep rows with overlapping genes\n",
    "\n",
    "# Create a boolean mask to filter rows based on the selected genes\n",
    "CosMx_mask_r = CosMx_transcripts_r['target'].isin(overlappinggenes_list)\n",
    "\n",
    "# Apply the mask to filter the DataFrame\n",
    "CosMx_transcripts_r_overlappinggenes = CosMx_transcripts_r[CosMx_mask_r]\n",
    "\n",
    "### Part 2: Quantify transcripts for overlapping genes df\n",
    "\n",
    "# Make new df\n",
    "CosMx_transcript_counts_per_cell_overlappinggenes_r = CosMx_transcripts_r_overlappinggenes.groupby('cell_id')['transcript_id'].nunique().reset_index()\n",
    "\n",
    "# Rename columns for clarity\n",
    "CosMx_transcript_counts_per_cell_overlappinggenes_r.columns = ['cell_id', 'transcript_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d8851da",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of transcripts per cell -- Overlapping genes\")\n",
    "print(f\"CosMx, Mean: {CosMx_transcript_counts_per_cell_overlappinggenes_r['transcript_count'].mean()}\")\n",
    "print(f\"CosMx, Median: {CosMx_transcript_counts_per_cell_overlappinggenes_r['transcript_count'].median()}\")\n",
    "print(f\"CosMx, Q1: {CosMx_transcript_counts_per_cell_overlappinggenes_r['transcript_count'].quantile(0.25)}\")\n",
    "print(f\"CosMx, Q3: {CosMx_transcript_counts_per_cell_overlappinggenes_r['transcript_count'].quantile(0.75)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe98e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Table with cell id and feature name values\n",
    "\n",
    "# Make df with just cell_id and feature_name\n",
    "CosMx_transcripts_r_featurespercell = CosMx_transcripts_r[['cell_id', 'target']]\n",
    "\n",
    "## Print unique features per cell\n",
    "\n",
    "## To answer the question: What unique features does each cell express?\n",
    "# Drop duplicate rows with same cell_id and feature_name combo\n",
    "CosMx_transcripts_r_uniquefeaturespercell = CosMx_transcripts_r_featurespercell.drop_duplicates(subset=['cell_id', 'target'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fea9ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantify the number of unique feature_name values per cell\n",
    "\n",
    "## To answer the question: How many unique features does each cell express\n",
    "\n",
    "# Make new df\n",
    "CosMx_transcripts_r_uniquefeaturespercell_count = CosMx_transcripts_r_uniquefeaturespercell.groupby('cell_id')['target'].nunique().reset_index()\n",
    "\n",
    "# Rename the new column for clarity\n",
    "CosMx_transcripts_r_uniquefeaturespercell_count.columns = ['cell_id', 'unique_features']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "500f5e17",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of unique features per cell -- All genes\")\n",
    "print(f\"CosMx, Mean: {CosMx_transcripts_r_uniquefeaturespercell_count['unique_features'].mean()}\")\n",
    "print(f\"CosMx, Median: {CosMx_transcripts_r_uniquefeaturespercell_count['unique_features'].median()}\")\n",
    "print(f\"CosMx, Q1: {CosMx_transcripts_r_uniquefeaturespercell_count['unique_features'].quantile(0.25)}\")\n",
    "print(f\"CosMx, Q3: {CosMx_transcripts_r_uniquefeaturespercell_count['unique_features'].quantile(0.75)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5760a16c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## What about for only overlapping genes?\n",
    "\n",
    "## To answer the question: How many unique features (of overlapping genes) does each cell have?\n",
    "\n",
    "\n",
    "### Part 1: Get rid of duplicate feature rows\n",
    "\n",
    "# Drop duplicate rows with same cell_id and feature_name combo\n",
    "CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes = CosMx_transcripts_r_overlappinggenes.drop_duplicates(subset=['cell_id', 'target'])\n",
    "\n",
    "### Part 2: Quantify unique features for overlapping genes df\n",
    "\n",
    "# Make new df\n",
    "CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes_count = CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes.groupby('cell_id')['target'].nunique().reset_index()\n",
    "\n",
    "# Rename the new column for clarity\n",
    "CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes_count.columns = ['cell_id', 'unique_features']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4bffb64",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of unique features per cell -- Overlapping genes\")\n",
    "print(f\"CosMx, Mean: {CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].mean()}\")\n",
    "print(f\"CosMx, Median: {CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].median()}\")\n",
    "print(f\"CosMx, Q1: {CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].quantile(0.25)}\")\n",
    "print(f\"CosMx, Q3: {CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].quantile(0.75)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfda0182",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Quantify the frequency / count that each feature appears in each cell (gene probe expression level)\n",
    "\n",
    "## To answer the question: What is the expression level of each feature in each cell? (Per gene, per cell)\n",
    "\n",
    "# Make a copy\n",
    "CosMx_transcripts_r_quantityofeachfeaturepercell = CosMx_transcripts_r_featurespercell.copy()\n",
    "\n",
    "# Quantify feature appearances\n",
    "CosMx_grouped_counts = CosMx_transcripts_r_quantityofeachfeaturepercell.groupby(['cell_id', 'target'])['target'].transform('count')\n",
    "CosMx_transcripts_r_quantityofeachfeaturepercell.loc[:, 'feature_quantity'] = CosMx_grouped_counts\n",
    "\n",
    "# Drop duplicate rows to keep only unique combinations\n",
    "CosMx_transcripts_r_quantityofeachfeaturepercell = CosMx_transcripts_r_quantityofeachfeaturepercell.drop_duplicates(subset=['cell_id', 'target'])\n",
    "\n",
    "# Resetting index for clarity\n",
    "CosMx_transcripts_r_quantityofeachfeaturepercell.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce6a0c1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## What about for only the overlapping genes?\n",
    "\n",
    "## To answer the question: What is the expression level of each overlapping feature in each cell? (Per gene, per cell)\n",
    "\n",
    "## Filter df to only keep rows with overlapping genes\n",
    "\n",
    "# Create a boolean mask to filter rows based on the selected genes\n",
    "CosMx_mask_r = CosMx_transcripts_r_quantityofeachfeaturepercell['target'].isin(overlappinggenes_list)\n",
    "\n",
    "# Apply the mask to filter the DataFrame\n",
    "CosMx_transcripts_r_quantityofeachfeaturepercell_overlappinggenes = CosMx_transcripts_r_quantityofeachfeaturepercell[CosMx_mask_r]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97ad21b8",
   "metadata": {},
   "source": [
    "#### Final set up for plots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24914864",
   "metadata": {},
   "source": [
    "CosMx vs Xenium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25e81028",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### CosMx df edits\n",
    "\n",
    "# Create a new DataFrame without the 'cell_id' column\n",
    "CosMx_r_ntrancriptsperfeature = CosMx_transcripts_r_quantityofeachfeaturepercell.drop('cell_id', axis=1).copy()\n",
    "\n",
    "# Group by 'target' and calculate the sum of 'feature_quantity'\n",
    "CosMx_r_ntrancriptsperfeature = CosMx_r_ntrancriptsperfeature.groupby('target')['feature_quantity'].mean().reset_index()\n",
    "\n",
    "# Rename the 'target' column to 'feature_name' to match Xenium\n",
    "CosMx_r_ntrancriptsperfeature = CosMx_r_ntrancriptsperfeature.rename(columns={'target': 'feature_name'})\n",
    "\n",
    "# Rename the 'feature_quantity' column to 'Xenium_meancounts' to match Xenium\n",
    "CosMx_r_ntrancriptsperfeature = CosMx_r_ntrancriptsperfeature.rename(columns={'feature_quantity': 'CosMx_r_meancounts'})\n",
    "\n",
    "# Display df\n",
    "display(CosMx_r_ntrancriptsperfeature)\n",
    "\n",
    "\n",
    "### Xenium df edits\n",
    "\n",
    "# Create a new DataFrame without the 'cell_id' column\n",
    "Xenium_r_ntrancriptsperfeature = Xenium_transcripts_r_quantityofeachfeaturepercell.drop('cell_id', axis=1).copy()\n",
    "\n",
    "# Group by 'feature_name' and calculate the sum of 'feature_quantity'\n",
    "Xenium_r_ntrancriptsperfeature = Xenium_r_ntrancriptsperfeature.groupby('feature_name')['feature_quantity'].mean().reset_index()\n",
    "\n",
    "# Rename the 'feature_quantity' column to 'Xenium_meancounts' to match Xenium\n",
    "Xenium_r_ntrancriptsperfeature = Xenium_r_ntrancriptsperfeature.rename(columns={'feature_quantity': 'Xenium_r_meancounts'})\n",
    "\n",
    "# Display df\n",
    "display(Xenium_r_ntrancriptsperfeature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "832bb27d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### All genes\n",
    "\n",
    "# Perform an outer merge based on the 'feature_name' column to keep all of the rows from both data frames\n",
    "merged_meancounts_allgenes = pd.merge(CosMx_r_ntrancriptsperfeature, Xenium_r_ntrancriptsperfeature, on='feature_name', how='outer')\n",
    "\n",
    "# Display\n",
    "display(merged_meancounts_allgenes)\n",
    "# 1131 rows (which includes all genes from CosMx and Xenium)\n",
    "\n",
    "\n",
    "### Overlapping genes\n",
    "\n",
    "# Perform an inner merge based on the 'feature_name' column\n",
    "merged_meancounts_overlappinggenes = pd.merge(CosMx_r_ntrancriptsperfeature, Xenium_r_ntrancriptsperfeature, on='feature_name', how='inner')\n",
    "\n",
    "# Display\n",
    "display(merged_meancounts_overlappinggenes)\n",
    "# 159 overlapping genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3cd86c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Expression of each gene within cells \\n\")\n",
    "\n",
    "print(\"Mean\")\n",
    "print(f\"Xenium, all genes: {merged_meancounts_allgenes['Xenium_r_meancounts'].mean(skipna=True)}\")\n",
    "print(f\"CosMx, all genes: {merged_meancounts_allgenes['CosMx_r_meancounts'].mean(skipna=True)}\")\n",
    "print(f\"Xenium, overlapping genes: {merged_meancounts_overlappinggenes['Xenium_r_meancounts'].mean()}\")\n",
    "print(f\"CosMx, overlapping genes: {merged_meancounts_overlappinggenes['CosMx_r_meancounts'].mean()}\")\n",
    "\n",
    "print(\" \")\n",
    "print(\"Median\")\n",
    "print(f\"Xenium, all genes: {merged_meancounts_allgenes['Xenium_r_meancounts'].median(skipna=True)}\")\n",
    "print(f\"CosMx, all genes: {merged_meancounts_allgenes['CosMx_r_meancounts'].median(skipna=True)}\")\n",
    "print(f\"Xenium, overlapping genes: {merged_meancounts_overlappinggenes['Xenium_r_meancounts'].median()}\")\n",
    "print(f\"CosMx, overlapping genes: {merged_meancounts_overlappinggenes['CosMx_r_meancounts'].median()}\")\n",
    "\n",
    "print(\" \")\n",
    "print(\"Q1\")\n",
    "print(f\"Xenium, all genes: {merged_meancounts_allgenes['Xenium_r_meancounts'].quantile(0.25)}\")\n",
    "print(f\"CosMx, all genes: {merged_meancounts_allgenes['CosMx_r_meancounts'].quantile(0.25)}\")\n",
    "print(f\"Xenium, overlapping genes: {merged_meancounts_overlappinggenes['Xenium_r_meancounts'].quantile(0.25)}\")\n",
    "print(f\"CosMx, overlapping genes: {merged_meancounts_overlappinggenes['CosMx_r_meancounts'].quantile(0.25)}\")\n",
    "\n",
    "print(\" \")\n",
    "print(\"Q3\")\n",
    "print(f\"Xenium, all genes: {merged_meancounts_allgenes['Xenium_r_meancounts'].quantile(0.75)}\")\n",
    "print(f\"CosMx, all genes: {merged_meancounts_allgenes['CosMx_r_meancounts'].quantile(0.75)}\")\n",
    "print(f\"Xenium, overlapping genes: {merged_meancounts_overlappinggenes['Xenium_r_meancounts'].quantile(0.75)}\")\n",
    "print(f\"CosMx, overlapping genes: {merged_meancounts_overlappinggenes['CosMx_r_meancounts'].quantile(0.75)}\")\n",
    "\n",
    "# Will ignore NA vales by default"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbb37cdd",
   "metadata": {},
   "source": [
    "Rep1 vs Rep2 Xenium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26a4179c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### Xenium (Rep1) df edits\n",
    "\n",
    "# Make copy\n",
    "Rep1_Xenium_transcripts_r_quantityofeachfeaturepercell = Xenium_transcripts_r_quantityofeachfeaturepercell.copy()\n",
    "\n",
    "# Create a new DataFrame without the 'cell_id' column\n",
    "Rep1_Xenium_r_ntrancriptsperfeature = Rep1_Xenium_transcripts_r_quantityofeachfeaturepercell.drop('cell_id', axis=1).copy()\n",
    "\n",
    "# Group by 'feature_name' and calculate the sum of 'feature_quantity'\n",
    "Rep1_Xenium_r_ntrancriptsperfeature = Rep1_Xenium_r_ntrancriptsperfeature.groupby('feature_name')['feature_quantity'].mean().reset_index()\n",
    "\n",
    "# Rename the 'feature_quantity' column to 'Xenium_meancounts' to match Xenium\n",
    "Rep1_Xenium_r_ntrancriptsperfeature = Rep1_Xenium_r_ntrancriptsperfeature.rename(columns={'feature_quantity': 'Rep1_Xenium_r_meancounts'})\n",
    "\n",
    "# Display df\n",
    "display(Rep1_Xenium_r_ntrancriptsperfeature)\n",
    "\n",
    "### Xenium (Rep2) df edits\n",
    "\n",
    "# Create a new DataFrame without the 'cell_id' column\n",
    "Rep2_Xenium_r_ntrancriptsperfeature = Rep2_Xenium_transcripts_r_quantityofeachfeaturepercell.drop('cell_id', axis=1).copy()\n",
    "\n",
    "# Group by 'feature_name' and calculate the sum of 'feature_quantity'\n",
    "Rep2_Xenium_r_ntrancriptsperfeature = Rep2_Xenium_r_ntrancriptsperfeature.groupby('feature_name')['feature_quantity'].mean().reset_index()\n",
    "\n",
    "# Rename the 'feature_quantity' column to 'Xenium_meancounts' to match Xenium\n",
    "Rep2_Xenium_r_ntrancriptsperfeature = Rep2_Xenium_r_ntrancriptsperfeature.rename(columns={'feature_quantity': 'Rep2_Xenium_r_meancounts'})\n",
    "\n",
    "# Display df\n",
    "display(Rep2_Xenium_r_ntrancriptsperfeature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2c1db94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform an outer merge based on the 'feature_name' column to keep all of the rows from both data frames\n",
    "XeniumReps_merged_meancounts = pd.merge(Rep1_Xenium_r_ntrancriptsperfeature, Rep2_Xenium_r_ntrancriptsperfeature, on='feature_name', how='outer')\n",
    "\n",
    "# Display\n",
    "display(XeniumReps_merged_meancounts)\n",
    "# 290 rows (all Xenium genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67cc235e",
   "metadata": {},
   "source": [
    "#### Output plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e221d28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatterplot for overlapping genes\n",
    "\n",
    "# Extract X and Y values from the DataFrame\n",
    "x_values = merged_meancounts_overlappinggenes['Xenium_r_meancounts']\n",
    "y_values = merged_meancounts_overlappinggenes['CosMx_r_meancounts']\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(7, 7))  # Width=7 inches, Height=7 inches\n",
    "\n",
    "# Calculate Pearson's correlation\n",
    "corr, p_value = pearsonr(x_values, y_values)\n",
    "\n",
    "# Calculate medians\n",
    "median_x = np.median(x_values)\n",
    "median_y = np.median(y_values)\n",
    "\n",
    "# Calculate median fold change -- using x / y to emphasize how the x values are larger\n",
    "median_fold_change = median_x / median_y if median_x != 0 else 0\n",
    "\n",
    "# Create plot \n",
    "plt.scatter(x_values, y_values)\n",
    "\n",
    "# Set axis limits (specify the range for the x and y axes)\n",
    "plt.xlim(0, 11)\n",
    "plt.ylim(0, 11) \n",
    "\n",
    "# Set custom tick locations and labels for the x and y axes\n",
    "# Customize the tick locations (as integers)\n",
    "plt.xticks([0, 2, 4, 6, 8, 10], fontsize=11)\n",
    "plt.yticks([0, 2, 4, 6, 8, 10], fontsize=11)  \n",
    "\n",
    "\n",
    "# Add labels next to each point using marker_styles and a value threshold (e.g., 1.0)\n",
    "value_threshold = 3.0  # Modify this threshold as needed\n",
    "\n",
    "annotations = []\n",
    "\n",
    "feature_names = merged_meancounts_overlappinggenes['feature_name']\n",
    "\n",
    "# Loop through the rows and label the points\n",
    "for i, feature_name in enumerate(feature_names):\n",
    "    x_val = x_values[i]\n",
    "    y_val = y_values[i]\n",
    "\n",
    "    if x_val > value_threshold or y_val > value_threshold:\n",
    "        label = plt.annotate(feature_name, (x_val, y_val), ha='left', fontsize=11)\n",
    "        annotations.append(label)\n",
    "\n",
    "# Use adjust_text to automatically adjust label positions\n",
    "adjust_text(annotations, arrowprops=dict(arrowstyle='->', color='black'))\n",
    "\n",
    "# Draw a dashed diagonal line through the plot\n",
    "plt.plot([0, 11], [0, 11], 'k--', linewidth=1) # 'k--' specifies a black, dashed line\n",
    "\n",
    "# Display Pearson's correlation coefficient, p-value, and median fold on the plot\n",
    "plt.text(0.5, 10.5, f'Pearson R = {corr:.2f}\\np value = {p_value:.2e}\\nMedian fold = {median_fold_change:.2f}', fontsize=12, ha='left', va='top', bbox=dict(boxstyle='round', facecolor='white', alpha=0.5))\n",
    "\n",
    "# Add labels and a title\n",
    "plt.xlabel('Xenium', fontsize=12)\n",
    "plt.ylabel('CosMx', fontsize=12)\n",
    "plt.title('Mean Transcript Counts (Overlapping Genes)', fontsize=12)\n",
    "\n",
    "# Disable the grid\n",
    "plt.grid(False)\n",
    "\n",
    "# Save plot\n",
    "#plt.savefig('/path/Scatterplot_MeanCounts_OverlappingGenesOnly.pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b02d543",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatterplot for all genes \n",
    "\n",
    "# Replace NaN values with -0.90\n",
    "merged_meancounts_allgenes.fillna(-0.90, inplace=True)\n",
    "\n",
    "# Axis tick code\n",
    "def format_tick_labels(axis_range):\n",
    "    # Generate tick labels: show labels for 0 and multiples of 2, empty string for others\n",
    "    labels = [str(i) if i % 2 == 0 or i == 0 else '' for i in axis_range]\n",
    "    return labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c94daa9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Continued from previous code block\n",
    "# Extract X and Y values from the DataFrame\n",
    "y_values = merged_meancounts_allgenes['CosMx_r_meancounts']\n",
    "x_values = merged_meancounts_allgenes['Xenium_r_meancounts']\n",
    "\n",
    "# Filter data to include only formerly NA points\n",
    "na_points = merged_meancounts_allgenes[(x_values == -0.90) | (y_values == -0.90)]\n",
    "na_x_values = na_points['Xenium_r_meancounts']\n",
    "na_y_values = na_points['CosMx_r_meancounts']\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(9, 13))  # Width=13 inches, Height=6 inches\n",
    "\n",
    "# Define colors based on the condition (darkorange for Xenium NA, cornflowerblue for CosMx NA)\n",
    "colors = ['darkorange' if x == -0.90 else 'cornflowerblue' for x, y in zip(na_x_values, na_y_values)]\n",
    "\n",
    "# Create plot \n",
    "plt.scatter(na_x_values, na_y_values, c=colors)\n",
    "\n",
    "# Set axis limits (specify the range for the x and y axes)\n",
    "plt.ylim(-1, 23)\n",
    "plt.xlim(-1, 12)\n",
    "\n",
    "# Set custom tick locations for the x and y axes\n",
    "x_ticks = range(-1, 12)\n",
    "y_ticks = range(-1, 23)  \n",
    "\n",
    "# Set custom tick labels, showing only 0 and multiples of 2\n",
    "plt.xticks(x_ticks, format_tick_labels(x_ticks), fontsize=11)\n",
    "plt.yticks(y_ticks, format_tick_labels(y_ticks), fontsize=11)\n",
    "\n",
    "# Add labels next to each point using marker_styles and a value threshold (e.g., 1.0)\n",
    "value_threshold = 3.0  # Modify this threshold as needed\n",
    "\n",
    "annotations = []\n",
    "\n",
    "feature_names = na_points['feature_name']\n",
    "\n",
    "# Loop through the rows and label the points\n",
    "for i, feature_name in enumerate(feature_names):\n",
    "    x_val = na_x_values.iloc[i]\n",
    "    y_val = na_y_values.iloc[i]\n",
    "\n",
    "    if x_val > value_threshold or y_val > value_threshold:\n",
    "        label = plt.annotate(feature_name, (x_val, y_val), ha='left', fontsize=11)\n",
    "        annotations.append(label)\n",
    "\n",
    "# Use adjust_text to automatically adjust label positions\n",
    "adjust_text(annotations, arrowprops=dict(arrowstyle='->', color='black'))\n",
    "\n",
    "# Add lines at 1, since 10x Genomics states that a high expression gene has >1 mean count per cell \n",
    "plt.axvline(x=1, color='gray', linestyle='--')\n",
    "plt.axhline(y=1, color='gray', linestyle='--')\n",
    "\n",
    "# Disable the grid\n",
    "plt.grid(False)\n",
    "\n",
    "# Add labels and a title\n",
    "plt.ylabel('CosMx', fontsize=12)\n",
    "plt.xlabel('Xenium', fontsize=12)\n",
    "plt.title('Mean Transcript Counts (Dataset-Specific Genes)', fontsize=12)\n",
    "\n",
    "# Save plot\n",
    "#plt.savefig('/path/Scatterplot_MeanCounts_DatasetSpecificGenes.pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09d78246",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatterplot for comparing Xenium reps\n",
    "\n",
    "# Extract X and Y values from the DataFrame\n",
    "x_values = XeniumReps_merged_meancounts['Rep1_Xenium_r_meancounts']\n",
    "y_values = XeniumReps_merged_meancounts['Rep2_Xenium_r_meancounts']\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(7, 7))  # Width=7 inches, Height=7 inches\n",
    "\n",
    "# Calculate Pearson's correlation\n",
    "corr, p_value = pearsonr(x_values, y_values)\n",
    "\n",
    "# Create plot \n",
    "plt.scatter(x_values, y_values)\n",
    "\n",
    "# Set axis limits (specify the range for the x and y axes)\n",
    "plt.xlim(0, 11)\n",
    "plt.ylim(0, 11) \n",
    "\n",
    "# Set custom tick locations and labels for the x and y axes\n",
    "# Customize the tick locations (as integers)\n",
    "plt.xticks([0, 2, 4, 6, 8, 10], fontsize=11)\n",
    "plt.yticks([0, 2, 4, 6, 8, 10], fontsize=11)  \n",
    "\n",
    "# Note: Removed the code to label the dots\n",
    "\n",
    "# Draw a dashed diagonal line through the plot\n",
    "plt.plot([0, 11], [0, 11], 'k--', linewidth=1) # 'k--' specifies a black, dashed line\n",
    "\n",
    "# Display Pearson's correlation coefficient and p-value on the plot\n",
    "plt.text(0.5, 10.5, f'Pearson R = {corr:.2f}\\np value = {p_value:.2e}', fontsize=12, ha='left', va='top', bbox=dict(boxstyle='round', facecolor='white', alpha=0.5))\n",
    "\n",
    "# Add labels and a title\n",
    "plt.xlabel('Xenium - Rep 1', fontsize=12)\n",
    "plt.ylabel('Xenium - Rep 2', fontsize=12)\n",
    "plt.title('Mean Transcript Counts (All Genes)', fontsize=12)\n",
    "\n",
    "# Disable the grid\n",
    "plt.grid(False)\n",
    "\n",
    "# Save plot\n",
    "#plt.savefig('/path/Scatterplot_XeniumReps_MeanCounts.pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31b866b9",
   "metadata": {},
   "source": [
    "#### Run MannWhitney tests (Wilcoxon) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7458825",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Transcript counts\n",
    "\n",
    "## All Genes\n",
    "\n",
    "# Create lists\n",
    "Xenium_transcript_counts_per_cell_stats = Xenium_transcript_counts_per_cell_r['transcript_count']\n",
    "CosMx_transcript_counts_per_cell_stats = CosMx_transcript_counts_per_cell_r['transcript_count']\n",
    "\n",
    "## Run MannWhitney test (Wilcoxon)\n",
    "res = stats.mannwhitneyu(Xenium_transcript_counts_per_cell_stats, \n",
    "                   CosMx_transcript_counts_per_cell_stats,\n",
    "                   use_continuity=True, \n",
    "                   alternative='two-sided', axis=0, method='auto')\n",
    "\n",
    "print(\"Transcript counts_All genes\")\n",
    "print(res)\n",
    "print(f'Statistic: {res.statistic}, p-value: {res.pvalue:.50f}')\n",
    "\n",
    "## Overlapping Genes\n",
    "\n",
    "# Create lists\n",
    "Xenium_transcript_counts_per_cell_overlappinggenes_stats = Xenium_transcript_counts_per_cell_overlappinggenes_r['transcript_count']\n",
    "CosMx_transcript_counts_per_cell_overlappinggenes_stats = CosMx_transcript_counts_per_cell_overlappinggenes_r['transcript_count']\n",
    "\n",
    "## Run MannWhitney test (Wilcoxon)\n",
    "res = stats.mannwhitneyu(Xenium_transcript_counts_per_cell_overlappinggenes_stats, \n",
    "                   CosMx_transcript_counts_per_cell_overlappinggenes_stats,\n",
    "                   use_continuity=True, \n",
    "                   alternative='two-sided', axis=0, method='auto')\n",
    "\n",
    "print(\" \")\n",
    "print(\"Transcript counts_Overlapping genes\")\n",
    "print(res)\n",
    "print(f'Statistic: {res.statistic}, p-value: {res.pvalue:.50f}')\n",
    "\n",
    "\n",
    "\n",
    "### Unique features\n",
    "\n",
    "## All Genes\n",
    "\n",
    "# Create lists\n",
    "Xenium_uniquefeatures_stats = Xenium_transcripts_r_uniquefeaturespercell_count['unique_features'].tolist()\n",
    "CosMx_uniquefeatures_stats = CosMx_transcripts_r_uniquefeaturespercell_count['unique_features'].tolist()\n",
    "\n",
    "## Run MannWhitney test (Wilcoxon)\n",
    "res = stats.mannwhitneyu(Xenium_uniquefeatures_stats, \n",
    "                   CosMx_uniquefeatures_stats,\n",
    "                   use_continuity=True, \n",
    "                   alternative='two-sided', axis=0, method='auto')\n",
    "\n",
    "print(\" \")\n",
    "print(\"Unique Features_All genes\")\n",
    "print(res)\n",
    "print(f'Statistic: {res.statistic}, p-value: {res.pvalue:.50f}')\n",
    "\n",
    "## Overlapping Genes\n",
    "\n",
    "# Create lists\n",
    "Xenium_uniquefeatures_overlappinggenes_stats = Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].tolist()\n",
    "CosMx_uniquefeatures_overlappinggenes_stats = CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].tolist()\n",
    "\n",
    "## Run MannWhitney test (Wilcoxon)\n",
    "res = stats.mannwhitneyu(Xenium_uniquefeatures_overlappinggenes_stats, \n",
    "                   CosMx_uniquefeatures_overlappinggenes_stats,\n",
    "                   use_continuity=True, \n",
    "                   alternative='two-sided', axis=0, method='auto')\n",
    "\n",
    "print(\" \")\n",
    "print(\"Unique Features_Overlapping genes\")\n",
    "print(res)\n",
    "print(f'Statistic: {res.statistic}, p-value: {res.pvalue:.50f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7fb7af0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Box and Whiskers Log scale transcript count plot for all and overlapping genes (with p-values)\n",
    "\n",
    "# Convert the 'transcript_count' columns to lists and combine them\n",
    "data = [\n",
    "    Xenium_transcript_counts_per_cell_r['transcript_count'].tolist(),\n",
    "    CosMx_transcript_counts_per_cell_r['transcript_count'].tolist(),\n",
    "    Xenium_transcript_counts_per_cell_overlappinggenes_r['transcript_count'],\n",
    "    CosMx_transcript_counts_per_cell_overlappinggenes_r['transcript_count']\n",
    "]\n",
    "\n",
    "# Custom labels\n",
    "labels = ['Xenium\\n ', 'CosMx\\n ', 'Xenium', 'CosMx']\n",
    "\n",
    "# Create a DataFrame from the lists for compatibility with seaborn\n",
    "df_data = pd.DataFrame(data).transpose()  # Transpose so each list becomes a column\n",
    "df_data.columns = labels  # Set the column names as your custom labels\n",
    "\n",
    "# Melt the DataFrame for plotting with seaborn\n",
    "df_long = df_data.melt(var_name='Technology', value_name='Number of Transcripts per Cell')\n",
    "\n",
    "# Now plotting with seaborn\n",
    "sns.set(style=\"ticks\", rc={\"figure.figsize\": (6, 6)})\n",
    "\n",
    "# Create a color palette\n",
    "palette = {\n",
    "    'Xenium\\n ': 'cornflowerblue',\n",
    "    'CosMx\\n ': 'darkorange',\n",
    "    'Xenium': 'cornflowerblue',\n",
    "    'CosMx': 'darkorange'\n",
    "}\n",
    "\n",
    "# Create the boxplot with seaborn\n",
    "sns.boxplot(data=df_long, x='Technology', y='Number of Transcripts per Cell', palette=palette, flierprops=dict(marker='o', markersize=4, color='black'))\n",
    "\n",
    "# Use log scale for the y-axis\n",
    "plt.yscale('log')  \n",
    "\n",
    "# Extend y-axis to 10000\n",
    "plt.ylim(1, 10000)\n",
    "\n",
    "# Manually adjusting the x-axis to add custom labels between certain ticks\n",
    "# There are 4 main categories plotted at positions 0, 1, 2, 3 on the x-axis\n",
    "tick_positions = [0, 1, 2, 3]  \n",
    "\n",
    "# Calculate midpoints for text placement\n",
    "midpoint_1 = (tick_positions[0] + tick_positions[1]) / 2\n",
    "midpoint_2 = (tick_positions[2] + tick_positions[3]) / 2\n",
    "\n",
    "# Adjust layout to make space for text below the x-axis\n",
    "plt.subplots_adjust(bottom=0.2)  # Increase bottom margin; adjust as needed\n",
    "\n",
    "# Get current axes\n",
    "ax = plt.gca()\n",
    "\n",
    "# Create a blended transformation\n",
    "# x in data coords, y in axes fraction coords\n",
    "transform = mtransforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "\n",
    "# Place custom text below the plot's y-axis using axes coordinates\n",
    "ax.text(midpoint_1, -0.1, 'All Genes', ha='center', va='top', fontsize=12, transform=transform)\n",
    "ax.text(midpoint_2, -0.1, 'Overlapping Genes', ha='center', va='top', fontsize=12, transform=transform)\n",
    "\n",
    "# Add stats\n",
    "ax.text(midpoint_1, 0.92, 'p<0.001', ha='center', va='top', fontsize=12, transform=transform)\n",
    "ax.text(midpoint_2, 0.92, 'p<0.001', ha='center', va='top', fontsize=12, transform=transform)\n",
    "\n",
    "plt.xlabel('', fontsize=12)\n",
    "plt.ylabel('Number of Transcripts per Cell', fontsize=12)\n",
    "plt.tick_params(axis='x', labelsize=12)\n",
    "plt.tick_params(axis='y', labelsize=11)\n",
    "sns.despine(offset=5, trim=True)\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save plot\n",
    "#plt.savefig('/path/BoxWhiskers_NumTranscripts_AllGenes_LogScale_WithPvalues_.pdf')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fd37516",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Box and Whiskers Log scale unique features plot for all and overlapping genes (with p-values)\n",
    "\n",
    "# Convert the 'unique_features' columns to lists and combine them\n",
    "data = [\n",
    "    Xenium_transcripts_r_uniquefeaturespercell_count['unique_features'].tolist(),\n",
    "    CosMx_transcripts_r_uniquefeaturespercell_count['unique_features'].tolist(),\n",
    "    Xenium_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].tolist(),\n",
    "    CosMx_transcripts_r_uniquefeaturespercell_overlappinggenes_count['unique_features'].tolist()\n",
    "]\n",
    "\n",
    "# Custom labels\n",
    "labels = ['Xenium\\n ', 'CosMx\\n ', 'Xenium', 'CosMx']\n",
    "\n",
    "# Create a DataFrame from the lists for compatibility with seaborn\n",
    "df_data = pd.DataFrame(data).transpose()  # Transpose so each list becomes a column\n",
    "df_data.columns = labels  # Set the column names as your custom labels\n",
    "\n",
    "# Melt the DataFrame for plotting with seaborn\n",
    "df_long = df_data.melt(var_name='Technology', value_name='Number of Unique Features per Cell')\n",
    "\n",
    "# Now plotting with seaborn\n",
    "sns.set(style=\"ticks\", rc={\"figure.figsize\": (6, 6)})\n",
    "\n",
    "# Create a color palette\n",
    "palette = {\n",
    "    'Xenium\\n ': 'cornflowerblue',\n",
    "    'CosMx\\n ': 'darkorange',\n",
    "    'Xenium': 'cornflowerblue',\n",
    "    'CosMx': 'darkorange'\n",
    "}\n",
    "\n",
    "# Create the boxplot with seaborn\n",
    "sns.boxplot(data=df_long, x='Technology', y='Number of Unique Features per Cell', palette=palette, flierprops=dict(marker='o', markersize=4, color='black'))\n",
    "\n",
    "# Use log scale for y-axis\n",
    "plt.yscale('log')\n",
    "\n",
    "# Extend y-axis to 1000\n",
    "plt.ylim(1, 1000)\n",
    "\n",
    "# Manually adjusting the x-axis to add custom labels between certain ticks\n",
    "# There are 4 main categories plotted at positions 0, 1, 2, 3 on the x-axis\n",
    "tick_positions = [0, 1, 2, 3]  \n",
    "\n",
    "# Calculate midpoints for text placement\n",
    "midpoint_1 = (tick_positions[0] + tick_positions[1]) / 2\n",
    "midpoint_2 = (tick_positions[2] + tick_positions[3]) / 2\n",
    "\n",
    "# Adjust layout to make space for text below the x-axis\n",
    "plt.subplots_adjust(bottom=0.2)  # Increase bottom margin; adjust as needed\n",
    "\n",
    "# Get current axes\n",
    "ax = plt.gca()\n",
    "\n",
    "# Create a blended transformation\n",
    "# x in data coords, y in axes fraction coords\n",
    "transform = mtransforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "\n",
    "# Place custom text below the plot's y-axis using axes coordinates\n",
    "ax.text(midpoint_1, -0.1, 'All Genes', ha='center', va='top', fontsize=12, transform=transform)\n",
    "ax.text(midpoint_2, -0.1, 'Overlapping Genes', ha='center', va='top', fontsize=12, transform=transform)\n",
    "\n",
    "# Add stats\n",
    "ax.text(midpoint_1, 0.92, 'p<0.001', ha='center', va='top', fontsize=12, transform=transform)\n",
    "ax.text(midpoint_2, 0.92, 'p<0.001', ha='center', va='top', fontsize=12, transform=transform)\n",
    "\n",
    "plt.xlabel('', fontsize=12)\n",
    "plt.ylabel('Number of Unique Features per Cell', fontsize=12)\n",
    "plt.tick_params(axis='x', labelsize=12)\n",
    "plt.tick_params(axis='y', labelsize=11)\n",
    "sns.despine(offset=5, trim=True)\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save plot\n",
    "#plt.savefig('/path/BoxWhiskers_NumUniqueFeatures_AllGenes_LogScale_WithPvalues_.pdf')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b1f67fa",
   "metadata": {},
   "source": [
    "## Cell-Related Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29677c6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in data\n",
    "\n",
    "Int_XeniumData = sc.read_h5ad('/path/25_11_22_Xenium_Dataset1_290_IntReps1and2_Annotated.h5ad')\n",
    "CosMxData = sc.read_h5ad('/path/25_11_22_CosMx_Annotated.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d2c1079-cd64-4a7b-802b-16fd1f0f735b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Subset Int_XeniumData to create a Rep1 only version\n",
    "\n",
    "Rep1_XeniumData = Int_XeniumData[Int_XeniumData.obs['batch'] == 'Sept'].copy()\n",
    "\n",
    "Rep1_XeniumData.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e898c08b",
   "metadata": {},
   "source": [
    "#### Cell area"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceb1d519",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create lists\n",
    "Xenium_cell_area_stats = Rep1_XeniumData.obs['cell_area_um2']\n",
    "CosMx_cell_area_stats = CosMxData.obs['cell_area_um2']\n",
    "\n",
    "## Run MannWhitney test (Wilcoxon)\n",
    "res = stats.mannwhitneyu(Xenium_cell_area_stats, \n",
    "                   CosMx_cell_area_stats,\n",
    "                   use_continuity=True, \n",
    "                   alternative='two-sided', axis=0, method='auto')\n",
    "\n",
    "print(\"Cell area\")\n",
    "print(res)\n",
    "print(f'Statistic: {res.statistic}, p-value: {res.pvalue:.5f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "638b3b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Box and Whiskers plot for cell area\n",
    "\n",
    "# Convert the 'cell_area_um2' columns to lists and combine them\n",
    "data = [\n",
    "    Rep1_XeniumData.obs[\"cell_area_um2\"].tolist(),\n",
    "    CosMxData.obs[\"cell_area_um2\"].tolist()\n",
    "]\n",
    "\n",
    "# Custom labels\n",
    "labels = ['Xenium', 'CosMx']\n",
    "\n",
    "# Create a DataFrame from the lists for compatibility with seaborn\n",
    "df_data = pd.DataFrame(data).transpose()  # Transpose so each list becomes a column\n",
    "df_data.columns = labels  # Set the column names as your custom labels\n",
    "\n",
    "# Melt the DataFrame for plotting with seaborn\n",
    "df_long = df_data.melt(var_name='Technology', value_name='Cell Area (um2)')\n",
    "\n",
    "# Now plotting with seaborn\n",
    "sns.set(style=\"ticks\", rc={\"figure.figsize\": (4.5, 6)})\n",
    "\n",
    "# Create a color palette\n",
    "palette = {\n",
    "    'Xenium': 'cornflowerblue',\n",
    "    'CosMx': 'darkorange'\n",
    "}\n",
    "\n",
    "# Create the boxplot with seaborn\n",
    "sns.boxplot(data=df_long, x='Technology', y='Cell Area (um2)', palette=palette, flierprops=dict(marker='o', markersize=4, color='black'))\n",
    "\n",
    "# Extend y-axis to 2000\n",
    "plt.ylim(0, 2000)\n",
    "\n",
    "# Specify custom y-axis ticks\n",
    "plt.yticks([0, 500, 1000, 1500, 2000])\n",
    "\n",
    "# Manually adjusting the x-axis to add custom labels between certain ticks\n",
    "# There are 4 main categories plotted at positions 0, 1, 2, 3 on the x-axis\n",
    "tick_positions = [0, 1, 2, 3]  \n",
    "\n",
    "# Calculate midpoints for text placement\n",
    "midpoint_1 = (tick_positions[0] + tick_positions[1]) / 2\n",
    "\n",
    "# Adjust layout to make space for text below the x-axis\n",
    "plt.subplots_adjust(bottom=0.2)  # Increase bottom margin; adjust as needed\n",
    "\n",
    "# Get current axes\n",
    "ax = plt.gca()\n",
    "\n",
    "# Create a blended transformation\n",
    "# x in data coords, y in axes fraction coords\n",
    "transform = mtransforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "\n",
    "# Place custom text using axes coordinates\n",
    "# Add stats\n",
    "ax.text(midpoint_1, 0.92, 'p=0.377', ha='center', va='top', fontsize=12, transform=transform)\n",
    "\n",
    "# Customize the plot\n",
    "plt.ylabel(\"Cell Area (um2)\", fontsize=12)\n",
    "plt.xlabel(\"Technology\", fontsize=12)\n",
    "#plt.title(\"Cell Area (um2)\", fontsize=12)\n",
    "sns.despine(offset=5, trim=True)  # Removes the top and right border spines\n",
    "\n",
    "plt.tick_params(axis='x', labelsize=12)\n",
    "plt.tick_params(axis='y', labelsize=11)\n",
    "\n",
    "# Adjust the plot\n",
    "plt.tight_layout()  # Adjust layout to make room for the legend if necessary\n",
    "\n",
    "# Save plot\n",
    "#plt.savefig('/path/BoxWhiskers_CellArea.pdf')\n",
    "\n",
    "# Show the figure\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "269eaf9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Cell area \\n\")\n",
    "\n",
    "print(\"Xenium\")\n",
    "print(f\"Mean: {Rep1_XeniumData.obs['cell_area_um2'].mean()}\")\n",
    "print(f\"Median: {Rep1_XeniumData.obs['cell_area_um2'].median()}\")\n",
    "print(f\"Q1: {Rep1_XeniumData.obs['cell_area_um2'].quantile(0.25)}\")\n",
    "print(f\"Q3: {Rep1_XeniumData.obs['cell_area_um2'].quantile(0.75)}\")\n",
    "print(f\"Maximum: {Rep1_XeniumData.obs['cell_area_um2'].max()}\")\n",
    "\n",
    "print(\" \")\n",
    "print(\"CosMx\")\n",
    "print(f\"Mean: {CosMxData.obs['cell_area_um2'].mean()}\")\n",
    "print(f\"Median: {CosMxData.obs['cell_area_um2'].median()}\")\n",
    "print(f\"Q1: {CosMxData.obs['cell_area_um2'].quantile(0.25)}\")\n",
    "print(f\"Q3: {CosMxData.obs['cell_area_um2'].quantile(0.75)}\")\n",
    "print(f\"Maximum: {CosMxData.obs['cell_area_um2'].max()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c22da7d",
   "metadata": {},
   "source": [
    "#### Transcript cellular location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8575441",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Stacked barplot for transcript cellular location\n",
    "\n",
    "## Note: CosMx is further broken down by Membrane and Cytoplasm, which isn't being specified here\n",
    "\n",
    "# Categories\n",
    "categories = ['Xenium', 'CosMx']\n",
    "\n",
    "# Numpy arrays\n",
    "Nuclear = [Xenium_r_overlaps_nucleus_1yes, CosMx_r_pct_nuc]\n",
    "NonNuclear = [Xenium_r_overlaps_nucleus_0no, (CosMx_r_pct_mem + CosMx_r_pct_cyt)]\n",
    "\n",
    "# Set the width of the bars\n",
    "bar_width = 0.35\n",
    "\n",
    "# Create a smaller figure and axis\n",
    "fig, ax = plt.subplots(figsize=(5, 6))  # Adjust the width and height as needed\n",
    "\n",
    "# Create stacked bar plot with reduced gap\n",
    "p1 = ax.bar(np.arange(len(categories)), Nuclear, bar_width, label='Nuclear', color='darkviolet')\n",
    "p2 = ax.bar(np.arange(len(categories)), NonNuclear, bar_width, bottom=Nuclear, label='Non-Nuclear', color='forestgreen')\n",
    "       \n",
    "# Labels\n",
    "for bar in ax.patches:\n",
    "    ax.text(bar.get_x() + bar.get_width() / 2,\n",
    "            bar.get_height() / 2 + bar.get_y(),\n",
    "            f'{round(bar.get_height())}%',  # Add \"%\" to the label\n",
    "            ha='center',\n",
    "            color='white',\n",
    "            size=13,\n",
    "            fontweight='bold')\n",
    "\n",
    "# Display legend\n",
    "ax.legend().set_visible(True)\n",
    "ax.legend(fontsize=11)\n",
    "\n",
    "# Add labels, title\n",
    "ax.set_ylabel('Percent of Transcripts in Dataset', fontsize=12)\n",
    "# ax.set_title('Transcript Cellular Composition', fontsize=11)\n",
    "\n",
    "# Set y-axis ticks and labels\n",
    "plt.tick_params(axis='y', labelsize=11)\n",
    "\n",
    "# Despine\n",
    "sns.despine(offset=5, trim=True)  # Removes the top and right border spines\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Set x-axis ticks and labels\n",
    "plt.xticks(ticks=np.arange(len(categories)), labels=categories, fontsize=12)\n",
    "\n",
    "# Save plot\n",
    "#plt.savefig('/path/StackedBarPlot_CellComp_Updated_.pdf')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25025fdf",
   "metadata": {},
   "source": [
    "#### Negative probes per cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72277fe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Negative probes per cell \\n\")\n",
    "\n",
    "print(\"Xenium\")\n",
    "print(f\"Mean: {Rep1_XeniumData.obs['control_probe_counts'].mean()}\")\n",
    "print(f\"Median: {Rep1_XeniumData.obs['control_probe_counts'].median()}\")\n",
    "print(f\"Q1: {Rep1_XeniumData.obs['control_probe_counts'].quantile(0.25)}\")\n",
    "print(f\"Q3: {Rep1_XeniumData.obs['control_probe_counts'].quantile(0.75)}\")\n",
    "print(f\"Maximum: {Rep1_XeniumData.obs['control_probe_counts'].max()}\")\n",
    "\n",
    "print(\" \")\n",
    "print(\"CosMx\")\n",
    "print(f\"Mean: {CosMxData.obs['nCount_negprobes'].mean()}\")\n",
    "print(f\"Median: {CosMxData.obs['nCount_negprobes'].median()}\")\n",
    "print(f\"Q1: {CosMxData.obs['nCount_negprobes'].quantile(0.25)}\")\n",
    "print(f\"Q3: {CosMxData.obs['nCount_negprobes'].quantile(0.75)}\")\n",
    "print(f\"Maximum: {CosMxData.obs['nCount_negprobes'].max()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00c59fbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_df = CosMxData.obs.sort_values(by='nCount_negprobes', ascending=False)\n",
    "\n",
    "display(sorted_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b5b319c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create lists\n",
    "Xenium_neg_probes_stats = Rep1_XeniumData.obs['control_probe_counts']\n",
    "CosMx_neg_probes_stats = CosMxData.obs['nCount_negprobes']\n",
    "\n",
    "## Run MannWhitney test (Wilcoxon)\n",
    "res = stats.mannwhitneyu(Xenium_neg_probes_stats, \n",
    "                   CosMx_neg_probes_stats,\n",
    "                   use_continuity=True, \n",
    "                   alternative='two-sided', axis=0, method='auto')\n",
    "\n",
    "print(\"Cell area\")\n",
    "print(res)\n",
    "print(f'Statistic: {res.statistic}, p-value: {res.pvalue:.50f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "520fcb80",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Box and Whiskers plot for negative probes per cell\n",
    "\n",
    "# Convert the associated negativate probe columns to lists and combine them\n",
    "data = [\n",
    "    Rep1_XeniumData.obs[\"control_probe_counts\"].tolist(),\n",
    "    CosMxData.obs[\"nCount_negprobes\"].tolist()\n",
    "]\n",
    "\n",
    "# Custom labels\n",
    "labels = ['Xenium', 'CosMx']\n",
    "\n",
    "# Create a DataFrame from the lists for compatibility with seaborn\n",
    "df_data = pd.DataFrame(data).transpose()  # Transpose so each list becomes a column\n",
    "df_data.columns = labels  # Set the column names as your custom labels\n",
    "\n",
    "# Melt the DataFrame for plotting with seaborn\n",
    "df_long = df_data.melt(var_name='Technology', value_name='Number of Negative Probes per Cell')\n",
    "\n",
    "# Now plotting with seaborn\n",
    "sns.set(style=\"ticks\", rc={\"figure.figsize\": (4.5, 6)})\n",
    "\n",
    "# Create a color palette\n",
    "palette = {\n",
    "    'Xenium': 'cornflowerblue',\n",
    "    'CosMx': 'darkorange'\n",
    "}\n",
    "\n",
    "# Create the boxplot with seaborn\n",
    "sns.boxplot(data=df_long, x='Technology', y='Number of Negative Probes per Cell', palette=palette, flierprops=dict(marker='o', markersize=4, color='black'))\n",
    "\n",
    "# Extend y-axis to 25\n",
    "plt.ylim(0, 25)\n",
    "\n",
    "# Specify custom y-axis ticks\n",
    "plt.yticks([0, 5, 10, 15, 20, 25])\n",
    "\n",
    "# Manually set the y-tick labels to only show particular values\n",
    "#plt.gca().set_yticklabels(['0', '', '10', '', '20', ''])\n",
    "\n",
    "# Manually adjusting the x-axis to add custom labels between certain ticks\n",
    "# There are 4 main categories plotted at positions 0, 1, 2, 3 on the x-axis\n",
    "tick_positions = [0, 1, 2, 3]  \n",
    "\n",
    "# Calculate midpoints for text placement\n",
    "midpoint_1 = (tick_positions[0] + tick_positions[1]) / 2\n",
    "\n",
    "# Adjust layout to make space for text below the x-axis\n",
    "plt.subplots_adjust(bottom=0.2)  # Increase bottom margin; adjust as needed\n",
    "\n",
    "# Get current axes\n",
    "ax = plt.gca()\n",
    "\n",
    "# Create a blended transformation\n",
    "# x in data coords, y in axes fraction coords\n",
    "transform = mtransforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "\n",
    "# Place custom text using axes coordinates\n",
    "# Add stats\n",
    "ax.text(midpoint_1, 0.92, 'p<0.001', ha='center', va='top', fontsize=12, transform=transform)\n",
    "\n",
    "# Customize the plot\n",
    "plt.ylabel(\"Number of Negative Probes per Cell\", fontsize=12)\n",
    "plt.xlabel(\"Technology\", fontsize=12)\n",
    "sns.despine(offset=5, trim=True)  # Removes the top and right border spines\n",
    "\n",
    "plt.tick_params(axis='x', labelsize=12)\n",
    "plt.tick_params(axis='y', labelsize=11)\n",
    "\n",
    "# Adjust the plot\n",
    "plt.tight_layout()  # Adjust layout to make room for the legend if necessary\n",
    "\n",
    "# Save plot\n",
    "#plt.savefig('/path/BoxWhiskers_NegProbesPerCell.pdf')\n",
    "\n",
    "# Show the figure\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21f0505b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "squidpy_plot",
   "language": "python",
   "name": "squidpy_plot"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
